{% extends 'layouts/base.html' %}

{% block title %}{{ config.MAIN_TITLE|safe }}{% endblock %}
{% block meta_description %}{% endblock %}
{% block head %}
    <link rel="stylesheet"
          href="//use.fontawesome.com/releases/v5.0.2/css/all.css">

    <link rel="stylesheet"
          href="//cdnjs.cloudflare.com/ajax/libs/animate.css/3.5.2/animate.min.css">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/CoreUI-v1.0.10/style.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/CoreUI-v1.0.10/simple-line-icons-2.4.1.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/please-wait-0.0.5/please-wait.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/bootstrap-multiselect-v0.9.13/bootstrap-multiselect.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/jquery-nstslider-1.0.13/dist/jquery.nstSlider.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/dynatables-0.3.1/jquery.dynatable.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/bootstrap-switch-3.3.4/css/bootstrap3/bootstrap-switch.min.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/bootstrap-hardskilled-extend-select-1.1.4/css/select.css', _external=True) }}">

    <link rel="stylesheet"
          href="{{ url_for('static', filename='styles/main.css', _external=True) }}">

{% endblock %}

{% block body %}
<body>

<div class="inner">

    <nav id="dataSetSelectMenu" class="navbar navbar-expand-sm bg-dark navbar-dark fixed-top">
        <a class="navbar-brand" href="#">{{ config.MAIN_TITLE|safe }}</a>
        <!-- will be filled in programatically //-->
    </nav>

    <div class="container-fluid">
        <main class="col-12" role="main">
            <div class="row equal">

                <!-- Gene/Protein/Pheno selection //-->

                <div class="col-md-6 col-lg-5 col-xl-4">
                    <div class="card h-100">
                        <div class="card-header">
                            <ul class="nav nav-tabs" role="tablist">
                                <li class="nav-item">
                                    <a class="nav-link" id="navPheno" data-toggle="tab" href="#tabSearchPheno"><i class="fas fa-search"></i> Phenotypes</a>
                                </li>
                                <li class="nav-item">
                                    <a class="nav-link" id="navGene" data-toggle="tab" href="#tabSearchGene"><i class="fas fa-search"></i> Search</a>
                                </li>
                                <li class="nav-item">
                                    <a class="nav-link" id="navLodPeaks" data-toggle="tab" href="#tabLodPeaks"><i class="fas fa-th"></i> LOD Peaks</a>
                                </li>
                            </ul>
                        </div>

                        <div class="card-body">
                            <div class="tab-content">
                                <div class="tab-pane h-100" role="tabpanel" id="tabSearchPheno">

                                    <div class="row">
                                        <div class="col" id="searchPheno">
                                        </div>
                                    </div>

                                    <div class="row row-spacer"></div>

                                    <div class="row">
                                        <div class="col" id="searchPhenoCategoryText">
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col" id="searchPhenoCategory">
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col">
                                            <div class="py-0" id="searchPhenoResultsInfo"></div>
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col results-wrapper">
                                            <div id="searchPhenoResultsDiv"></div>
                                        </div>
                                    </div>
                                </div>

                                <div class="tab-pane h-100" role="tabpanel" id="tabSearchGene">
                                    <div class="row">
                                        <div class="col">
                                            <div class="input-group">
                                                <input type="text" id="searchTerm" name="searchTerm" class="form-control" placeholder="Please type a term to search for...">
                                                <span class="input-group-btn">
                                                    <button id="btnGo" class="btn btn-primary"><i class="fas fa-search"></i></button>
                                                </span>
                                            </div>
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col">
                                            <div class="py-2" id="searchResultsTableInfo"></div>
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col results-wrapper">
                                            <div id="searchResultsDiv"></div>
                                        </div>
                                    </div>
                                </div>

                                <div class="tab-pane h-100" role="tabpanel" id="tabLodPeaks">
                                    Peaks
                                </div>
                            </div>
                        </div>
                    </div>
                </div>

                <!-- end Gene/Protein/Pheno selection //-->

                <!-- LOD Scan section //-->

                <div class="col-md-6 col-lg-7 col-xl-8">

                    <!-- only display at first, remove after a succesful selection //-->
                    <div class="row" id="divWelcomeMesage">
                        <div class="col">
                            <div class="jumbotron jumbotron-fluid" style="background: none">
                                <div class="container">
                                    <h1 class="display-3">{{ config.MAIN_TITLE|safe }}</h1>
                                    <p class="lead">Please search for a term of interest.</p>
                                </div>
                            </div>
                        </div>
                    </div>

                    <div class="row invisible" id="divItemInfo">
                        <div class="col">
                            <div class="card">
                                <div class="card-header">
                                    <a class="collapsed" data-toggle="collapse" href="#collapseExample" aria-expanded="false" aria-controls="collapseExample">
                                    <i class="fa" aria-hidden="true"></i>
                                    <span id="div_current_item_information_header"></span>
                                    </a>
                                </div>
                                <div class="card-body collapse" id="collapseExample">
                                    <div id="div_current_item_information_body"></div>
                                </div>
                            </div>
                        </div>
                    </div>

                    <div class="row invisible" id="divLOD">
                        <div class="col">
                            <div class="card h-100">
                                <div class="card-header" id="lod">
                                    <i class="fas fa-chart-area"></i> LOD Plot
                                </div>
                                <div class="card-body">
                                    <div class="row">
                                        <div class="col">
                                            <div id="lodPlotChart"></div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>

                <!-- end LOD Scan section //-->

            </div>

            <div class="row row-spacer"></div>

            <div class="row equal invisible" id="divSecondRow">

                <!-- Profile plot //-->

                <div class="col-md-6 col-lg-5 col-xl-4">
                    <div class="card h-100">
                        <div class="card-header" id="profilePlot">
                            <ul id="infoTabs" class="nav nav-tabs" role="tablist">
                                <li class="nav-item">
                                    <a class="nav-link active" id="navProfile" data-toggle="tab" href="#tabProfile"><i class="fas fa-calculator"></i> Profile Plot</a>
                                </li>
                                <li class="nav-item">
                                    <a class="nav-link" id="navCorrelation" data-toggle="tab" href="#tabCorrelation"><i class="fas fa-barcode"></i> Correlation</a>
                                </li>
                            </ul>
                        </div>
                        <div class="card-body">

                            <div class="tab-content">
                                <div role="tabpanel" class="tab-pane active" id="tabProfile">
                                    <div class="row">
                                        <div class="col font-weight-bold">
                                            Select your factors
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col-8">
                                            <div id="factor-view-select-wrapper">
                                                <select id="factor-view-select" data-style="btn-secondary" multiple="multiple">
                                                </select>
                                            </div>
                                        </div>
                                        <div class="col-4">
                                            <button id="btnFactPlot" type="button" class="btn btn-flat btn-primary">
                                                <i class="fas fa-sync"></i> Update</button>
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col font-weight-bold">
                                            Select a series to color
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col" id="divFactorSeries">
                                        </div>
                                    </div>

                                    <div class="row row-spacer"></div>
                                    <div class="row" id="fact-svg-panel">
                                        <div class="col">
                                            <div id="profilePlotChart"></div>
                                        </div>
                                    </div>
                                </div>
                                <div role="tabpanel" class="tab-pane" id="tabCorrelation">
                                    <div class="row">
                                        <div class="col">
                                            <span class="align-bottom">Select Correlation Dataset:</span>
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col" id="correlationDatasetSelection">
                                        </div>
                                    </div>

                                    <div class="row row-spacer"></div>

                                    <div class="row">
                                        <div class="col results-wrapper-225" id="correlationDataTable">
                                        </div>
                                    </div>

                                    <div class="row row-spacer"></div>

                                    <div class="row">
                                        <div class="col">
                                            <button id="btnDownloadCorrelationData" type="button" class="btn btn-sm btn-flat btn-primary">
                                                <i class="fas fa-download"></i> Download Correlations</button>
                                        </div>
                                    </div>

                                    <div class="row row-spacer"></div>

                                    <div class="row">
                                        <div class="col">
                                            Select an ID from above to plot the correlation data.
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col" id="correlationDataPlot">
                                            <div id="plotCorrelation"></div>
                                        </div>
                                    </div>

                                    <div class="row row-spacer"></div>

                                    <div class="row">
                                        <div class="col font-weight-bold">
                                            Select a series to color
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col" id="divFactorSeriesCorrelation">
                                        </div>
                                    </div>
                                </div>
                            </div>


                        </div>
                    </div>
                </div>

                <!-- end Profile plot //-->

                <!-- Secondary plots //-->

                <div class="col-md-6 col-lg-7 col-xl-8">
                    <div class="card" id="secondaryPlots">
                        <div class="card-header">
                            <ul id="secondaryPlotTabs" class="nav nav-tabs" role="tablist">
                                <li class="nav-item">
                                    <a class="nav-link active" id="navPlotEffect" data-toggle="tab" href="#tabPlotEffect"><i class="fas fa-chart-line"></i> Effect</a>
                                </li>
                                <li class="nav-item">
                                    <a class="nav-link" id="navPlotMediation" data-toggle="tab" href="#tabPlotMediation"><i class="far fa-calendar-alt fa-rotate-180"></i> Mediation</a>
                                </li>
                                <li class="nav-item">
                                    <a class="nav-link" id="navPlotSNPs" data-toggle="tab" href="#tabPlotSNPs"><i class="fas fa-signal fa-rotate-90"></i> SNP Association</a>
                                </li>
                            </ul>
                        </div>
                        <div class="card-body">
                            <div class="tab-content">
                                <div role="tabpanel" class="tab-pane active" id="tabPlotEffect">
                                    <div class="row">
                                        <div class="col">
                                            Best Linear Unbiased Predictors (BLUPs)
                                            <input id="performBLUP" name="performBLUP" type="checkbox" data-size="mini" data-on-color="info">
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col">
                                            <div id="plotEffect"></div>
                                        </div>
                                    </div>
                                </div>
                                <div role="tabpanel" class="tab-pane" id="tabPlotMediation">
                                    <div class="row">
                                        <div class="col" id="divMediationSelect">
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col">
                                            <div id="plotMediation"></div>
                                        </div>
                                    </div>
                                </div>
                                <div role="tabpanel" class="tab-pane" id="tabPlotSNPs">
                                    <div class="row" id="snpWindowMenu">
                                    </div>
                                    <div class="row">
                                        <div class="col">
                                            <div id="plotSNPAssoc" style="height:800px"></div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>

                <!-- end Secondary plots //-->

            </div>

        </main>
    </div>

<!-- make some nice circles //-->
<div id="ap" class="invisible">
<svg class="lds-curve-bars" width="200px"  height="200px"  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" style="background: rgba(0, 0, 0, 0) none repeat scroll 0% 0%;"><g transform="translate(50,50)">
</circle><circle cx="0" cy="0" r="18" fill="none" stroke="#1aafd0" stroke-width="1" stroke-dasharray="52 52">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur="1.4s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.2" repeatCount="indefinite"></animateTransform>
</circle><circle cx="0" cy="0" r="22" fill="none" stroke="#6a67ce" stroke-width="1" stroke-dasharray="78 78">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur=".8s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.4" repeatCount="indefinite"></animateTransform>
</circle><circle cx="0" cy="0" r="26" fill="none" stroke="#ffb900" stroke-width="1" stroke-dasharray="104 104">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur="1s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.6" repeatCount="indefinite"></animateTransform>
</circle><circle cx="0" cy="0" r="30" fill="none" stroke="#fc636b" stroke-width="1" stroke-dasharray="130 130">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur="1.2s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.8" repeatCount="indefinite"></animateTransform>
</circle></g></svg>
</div>

<!-- for determining font size //-->
<div class="invisible">
    <svg version="1.1" class="highcharts-root" style="font-family:-apple-system, system-ui, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;;font-size: 15px;" xmlns="http://www.w3.org/2000/svg" width="727" height="800" viewBox="0 0 727 800" class="invisible">
        <g class="highcharts-data-labels highcharts-series-1 highcharts-xrange-series highcharts-tracker " visibility="hidden">
            <g class="highcharts-label highcharts-data-label highcharts-data-label-color-0">
                <text id="getTheFontSize" x="5" style="font-size:11px;font-weight:bold;color:white;fill:white;" y="16">
                    <tspan id="val1" x="5" y="16" class="highcharts-text-outline" fill="#000000" stroke="#000000" stroke-width="2px" stroke-linejoin="round" style="">TEST</tspan>
                    <tspan id="val2" x="5" y="16">TEST</tspan>
                </text>
            </g>
        </g>
    </svg>
</div>


<!-- Main Footer -->
<nav class="navbar fixed-bottom navbar-dark bg-dark">
    <ul class="navbar-nav mr-auto">
      <li class="nav-item">
        <span class="text-muted">
            <strong>Created by <a href="http://churchill-lab.jax.org">The Churchill Lab</a>.</strong>
        </span>
      </li>
    </ul>
</nav>
<!-- end Main Footer -->


</div>
</body>
{% endblock %}

{% block javascript %}
    <!-- please wait -->
    <script src="{{ url_for('static', filename='vendor/please-wait-0.0.5/please-wait.js') }}"></script>

    <!-- Loading app //-->
    <script>
    window.loading_screen = window.pleaseWait({
        logo: '',
        backgroundColor: '#000000',
        loadingHtml: '<div id="loader-wrapper"><div id="loader"></div><div class="loader-section section-left"></div><div class="loader-section section-right"></div></div>'
    });
    </script>

    <script src="//code.jquery.com/jquery-3.2.1.min.js"
            integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4="
            crossorigin="anonymous"></script>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js"
            integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q"
            crossorigin="anonymous"></script>

    <script src="//maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js"
            integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl"
            crossorigin="anonymous"></script>

    <!-- ENSIMPL API -->
    <script src="//churchill-lab.jax.org/ensimpl/api/js/ensimpl.js"></script>

    <!-- Highcharts/Highstock //-->
    <script src="//code.highcharts.com/stock/6.1.0/highstock.src.js"></script>
    <script src="//code.highcharts.com/stock/6.1.0/highcharts-more.js"></script>
    <script src="//code.highcharts.com/stock/6.1.0/modules/xrange.js"></script>
    <script src="//code.highcharts.com/stock/6.1.0/modules/boost.js"></script>
    <script src="//code.highcharts.com/stock/6.1.0/modules/exporting.js"></script>
    <script src="//code.highcharts.com/stock/6.1.0/modules/offline-exporting.js"></script>

    <!-- D3 //-->
    <script src="//d3js.org/d3.v3.min.js"></script>
    <script src="//d3js.org/d3-queue.v3.min.js"></script>
    <script src="//d3js.org/d3-request.v1.min.js"></script>

    <script src="//cdnjs.cloudflare.com/ajax/libs/bootstrap-slider/10.0.0/bootstrap-slider.min.js"></script>

    <!-- Extra //-->
    <script src="{{ url_for('static', filename='vendor/CoreUI-v1.0.6/app.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/sweetalert2.all.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/js.cookie-2.2.0.min.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/seedrandom-2.4.3/seedrandom.min.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/bootstrap-multiselect-v0.9.13/bootstrap-multiselect.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/jquery-nstslider-1.0.13/dist/jquery.nstSlider.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/bootstrap-notify-3.1.3/bootstrap-notify.min.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/dynatables-0.3.1/jquery.dynatable.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/bootstrap-hardskilled-extend-select-1.1.4/js/select.min.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/bootstrap-switch-3.3.4/js/bootstrap-switch.min.js') }}"></script>
    <script src="{{ url_for('static', filename='js/utilities.js') }}"></script>

    <script>

    /**
     * Show a debug message on the screen.
     * @param {string} title - title for message
     * @param {string} message - informational message
     */
    function debugMessage(title, message) {
        {% if debug %}
        $.notify({
            title: title,
            message: message,
            icon: 'fas fa-user-secret',
        },{
            type: 'minimalist',
            delay: 2000,
            mouse_over: 'pause',
            template: '<div data-notify="container" class="col-xs-11 col-sm-3 alert alert-{0}" role="alert">' +
                '<button type="button" aria-hidden="true" class="close" data-notify="dismiss">×</button>' +
                '<div class="row"><div class="col"><span data-notify="icon"></span></col><col><span data-notify="title">{1}</span></div></div>' +
                '<div class="row"><div class="col"><span data-notify="message">{2}</span></div></div>' +
            '</div>'
        });
        {% endif %}
    }

    /**
     * Add a button to the Highcharts export menu.
     * @param {string} mytext - title for button
     * @param {function} myfunction - function to call
     * @return {Object} buttons for highchart export
     */
    function appendExportButton(mytext, myfunction){
        // get default Highchart Export buttons
        let defaultButtons = Highcharts.getOptions().exporting.buttons;
        let myButtons = $.extend(true, {}, defaultButtons);

        myButtons.contextButton.menuItems.push({
            text: mytext,
            onclick: myfunction
        });

        return {
            buttons: myButtons
        };
    }

    /**
     *  Extended disable function
     */
    jQuery.fn.extend({
        disable: function(state) {
            return this.each(function() {
                let $this = $(this);
                if($this.is('input, button, textarea, select'))
                    this.disabled = state;
                else
                    $this.toggleClass('disabled', state);
            });
        }
    });

    /**
     * Download data.
     *
     * From d3.js documentation:
     *
     * When a task completes, it must call the provided callback.
     *
     * The first argument to the callback should be null if the task is
     * successful, or the error if the task failed.
     *
     * The optional second argument to the callback is the return value of the task.
     * (To return multiple values from a single callback, wrap the results in an object or array.)
     *
     * @param {string} URL - the URL to download data
     * @param {string} description - descriptive text
     * @param {function} callback - the callback function
     */
    function downloadData(URL, description, callback) {
        console.log('Downloading: ', URL);

        $.ajax({
            url: URL,
            method: 'GET'
        }).done(function(data, textStatus, jqXHR) {
            console.log('Download of ' + description + ':', data);
            callback(null, data);
        }).fail(function(jqXHR, textStatus, errorThrown) {
            //displayError('Server Error', 'Unable to download ' + description);
            console.error(description, textStatus);
            callback(errorThrown, null);
        });
    }

    QTL = function() {
        var g = {};

        function init() {
            g.dataSetID = null;
            g.dataSets = {};
            g.numMarkers = null;
            g.ensemblVersion = null;
            // TODO: do not hard code species
            g.speciesID = 'Mm';
            g.chromosomes = null;
            g.runningTask = false;
            g.chartLOD = null;
            g.chartMediation = null;
            g.chartEffect = null;
            g.chartSNPAssoc = null;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;

            //
            g.geneID = null;
            g.proteinID = null;
            g.phenotypeID = null;
            g.ensimpl = null;
            g.LOD = null;
            g.lodPeaks = null;
            g.chartPeaks = null;
            g.phenoDynaTable = null;
            g.expressionData = null;
            g.searchTable = null;
            g.orderCount = null;
            g.currentGene = null;
            g.fact = null;

            d3.queue()
                .defer(downloadDataSets)
                .await(function(error, data) {
                    console.log('dataSetsData=', data);

                    if (error) {
                        console.error(error);
                        {% if not debug -%}
                        // ?debug=1 will prevent forwarding on error
                        window.location = '{{ url_for('page.error') }}';
                        {%- endif %}
                        return;
                    }

                    if ((data.result === undefined) || (data.result.dataSets.length === 0)) {
                        console.error('data=', data);
                        {% if not debug -%}
                        // ?debug=1 will prevent forwarding on error
                        window.location = '{{ url_for('page.error') }}';
                        {%- endif %}
                        return;
                    }

                    g.dataSets = {};
                    g.numMarkers = data.result.numMarkers;
                    g.ensemblVersion = data.result.ensemblVersion;

                    console.log(data.result.dataSets);

                    $.each(data.result.dataSets, function(dsi, ds) {
                        ds.gene_ids = {};
                        ds.protein_ids = {};
                        ds.phenotypes = {};

                        if (ds.dataType === 'mRNA') {
                            // convert array to object (hash)
                            $.each(ds.annotations.ids, function(index, element) {
                                ds.gene_ids[element] = 1;
                            });

                        } else if (ds.dataType === 'protein') {
                            $.each(ds.annotations.ids, function(index, element) {
                                if (!(element.gene_id in ds.gene_ids)) {
                                    ds.gene_ids[element.gene_id] = {
                                        'id': element.gene_id,
                                        'protein_ids':[]
                                    }
                                }
                                ds.protein_ids[element.protein_id] = element.gene_id;
                                ds.gene_ids[element.gene_id]['protein_ids'].push(element.protein_id);
                            });

                        } else if (ds.dataType === 'phenotype') {
                            $.each(ds.annotations, function(index, element) {
                                ds.phenotypes[element.dataName] = element;
                            });
                        } else {
                            // TODO: handle error
                            console.error('ERROR in dataSet data:', ds);
                        }

                        // don't need the extra annotations laying around
                        delete ds['annotations'];

                        g.dataSets[ds.id] = ds;
                    });

                    let requestedDataSetID = '{{ datasetid }}';

                    if (g.dataSets[requestedDataSetID] !== undefined) {
                        g.dataSetID = requestedDataSetID;
                    } else {
                        g.dataSetID = data.result.dataSets[0].id;
                    }

                    if (data.result.dataSets.length === 1) {
                        // just show a menu item
                        let html = `<ul class="navbar-nav navbar-dark mr-auto">
                                        <li class="nav-item active" style="color:white">
                                            ${g.dataSets[g.dataSetID].displayName}
                                        </li>
                                    </ul>`;

                        $('#dataSetSelectMenu').append(html);
                        $('#correlationDatasetSelection').html(`${g.dataSets[g.dataSetID].displayName}`);
                        $('#correlationDatasetSelection').append(`<input type="hidden" id="correlationDatasetSelect" value="${g.dataSets[g.dataSetID].id}"/>`);

                    } else if (data.result.dataSets.length > 1) {
                        // show multiple menu items
                        let html = `<button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarTogglerQTL" aria-controls="navbarTogglerQTL" aria-expanded="false" aria-label="Toggle navigation">
                                        <span class="navbar-toggler-icon"></span>
                                    </button>
                                    <div class="collapse navbar-dark navbar-collapse" id="navbarTogglerQTL">
                                    <span class="navbar-text text-white pr-2">Current Data Set <i class="fas fa-long-arrow-alt-right"></i></span>
                                    <select id="datasetSelect" data-btn-class="btn-outline-secondary btn-sm" class="form-control">`;

                        for (let d in data.result.dataSets) {
                            html += `<option value="${data.result.dataSets[d].id}">${data.result.dataSets[d].displayName}</option>`;
                        }

                        html += '</select></div>';
                        $('#dataSetSelectMenu').append(html);
                        $('#datasetSelect').extendSelect();
                        $('#datasetSelect').on('change', function () {
                            let selected = $(this).find(':selected').val();
                            switchDataSet(selected);
                        });

                        ///

                        // show multiple menu items
                        let htmlNew = `<select id="correlationDatasetSelect" data-btn-class="btn-secondary btn-sm" class="form-control">`;

                        for (let d in data.result.dataSets) {
                            htmlNew += `<option value="${data.result.dataSets[d].id}">${data.result.dataSets[d].displayName}</option>`;
                        }

                        htmlNew += '</select>';

                        $('#correlationDatasetSelection').html(htmlNew);
                        //$('#correlationDatasetSelect').extendSelect();
                        $('#correlationDatasetSelect').on('change', function () {
                            let selected = $(this).find(':selected').val();
                            switchCorrelationDataSet(selected);
                            //console.log(selected);
                        });
                    }

                    configureCovarFactors();

                    $('#btnFactPlot').on('click', function(evt) {
                        if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                            plotProfile(g.geneID);
                        } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                            plotProfile(g.proteinID);
                        } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                            plotProfile(g.phenotypeID);
                        } else {
                            // TODO: major error
                            console.error('MAJOR ERROR');
                        }
                    });

                    $('#btnDownloadCorrelationData').on('click', function(evt) {
                        if (g.correlationData) {
                            let csvContent = '';
                            let id = '';

                            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                                id = g.geneID;
                            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                                id = g.proteinID;
                            } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                                id = g.phenotypeID;
                            } else {
                                console.error('MAJOR ERROR');
                            }

                            if (g.dataSets[g.correlationDatasetID].dataType === 'mRNA') {
                                csvContent = '"id","symbol","chr","start","end","correlation"\n';
                                $.each(g.correlationData, function (k, v) {
                                    csvContent += `"${v['id']}","${v['symbol']}","${v['chr']}",${v['start']},${v['end']},${v['cor']}\n`;
                                });
                            } else if (g.dataSets[g.correlationDatasetID].dataType === 'protein') {
                                csvContent = '"id","symbol","chr","start","end","correlation"\n';
                                $.each(g.correlationData, function (k, v) {
                                    csvContent += `"${v['id']}","${v['symbol']}","${v['chr']}",${v['start']},${v['end']},${v['cor']}\n`;
                                });
                            } else if (g.dataSets[g.correlationDatasetID].dataType === 'phenotype') {
                                csvContent = '"id","correlation"\n';
                                $.each(g.correlationData, function (k, v) {
                                    csvContent += `"${v['id']}",${v['cor']}\n`;
                                });
                            } else {
                                // TODO: major error
                                console.error('MAJOR ERROR');
                            }

                            downloadCSV(csvContent, `${id}_correlation.csv`, 'text/csv;encoding:utf-8');
                        } else {
                            console.log('no correlation data?');
                        }
                    });


                    g.ensimpl = new ensimpl({
                        search_limit: 100,
                        search_version: data.result.ensemblVersion,
                        search_species: g.speciesID
                    });


                    d3.queue()
                        .defer(downloadChromosomes, data.result.ensemblVersion, g.speciesID)
                        .defer(downloadLODPeaks, g.dataSetID)
                        .await(function(error, chromosomes, lodPeaks) {
                            g.chromosomes = {'idx': [], 'chr': {}};

                            let configChromosomes = {{ config.CHROMOSOMES|safe }};
                            let start = 1;
                            let end = 0;

                            $.each(chromosomes.chromosomes, function(index, element) {
                                if (configChromosomes.indexOf(element.chromosome) !== -1) {
                                    let chrom = element;

                                    start = end + 1;
                                    end += chrom.length;

                                    chrom.start = start;
                                    chrom.end = end;
                                    chrom.mid = chrom.start + Math.ceil((chrom.end - chrom.start) / 2);

                                    g.chromosomes['idx'].push(chrom);
                                    g.chromosomes['chr'][element.chromosome] = chrom;
                                }
                            });

                            g.chromosomes['totalLength'] = end;

                            setDataSet(g.dataSetID, true, lodPeaks.result);

                            window.loading_screen.finish();
                        });
                });
            }

        /**
         * Display the gene information.
         * @param {Object} geneData - gene information
         */
        function displayGeneData(geneData) {
            // TODO: link to database based upon ensembl version?

            console.log('Displaying gene information:', geneData);
            for(let key in geneData) {
                geneData = geneData[key];
            }

            let startFormat = formatMbp(geneData.start/1000000);
            let endFormat = formatMbp(geneData.end/1000000);
            let mbpDesc = 'Mbp';

            if (startFormat < 1) {
                startFormat = geneData.start;
                endFormat = geneData.end;
                mbpDesc = 'bp';
            }

            let strand = "forward";
            if (geneData.strand === "-") {
                strand = "negative";
            }

            let htmlHeader = geneData.id;
            htmlHeader += ' <em>' + geneData.symbol + '</em>';

            let htmlBody = '<table class="table table-sm">';
            htmlBody += '<tr><td class="border-0"><strong>ID</strong></td><td class="border-0">' + geneData.id;
            htmlBody += '<br>';
            htmlBody += '<small><a target="_newwin" href="http://www.ensembl.org/id/' + geneData.id + '">Ensembl <i class="fas fa-external-link-alt" aria-hidden="true"></i></small></a>';

            $.each(geneData.external_ids, function(idx, elem) {
                let url = '';

                if (elem.db === 'EntrezGene') {
                    url = 'https://www.ncbi.nlm.nih.gov/gene/?term=' + elem.db_id;
                } else if (elem.db === 'MGI') {
                    url = 'http://www.informatics.jax.org/marker/' + elem.db_id;
                } else if (elem.db === 'UniGene') {
                    url = 'https://www.ncbi.nlm.nih.gov/unigene/?term=' + elem.db_id;
                }

                if (url !== '') {
                    htmlBody += ' <small><a target="_newwin" href="' + url + '">' + elem.db + ' <i class="fas fa-external-link-alt" aria-hidden="true"></i></small></a>';
                }
            });

            htmlBody += '</td></tr>';
            htmlBody += `<tr><td><strong>Symbol</strong></td><td>${geneData.symbol}</td></tr>`;
            htmlBody += `<tr><td><strong>Location</strong></td><td>${geneData.chromosome}:${startFormat}-${endFormat}</td></tr>`;

            if (geneData.synonyms !== undefined) {
                let synonyms = geneData.synonyms.join(', ');
                htmlBody += `<tr><td><strong>Synonyms</strong></td><td>${synonyms}</td></tr>`;
            }

            if (geneData.name) {
                htmlBody += `<tr><td><strong>Description</strong></td><td>${geneData.name}</td></tr>`;
            }

            htmlBody += '</table>';

            $('#div_current_item_information_header').html(htmlHeader);
            $('#div_current_item_information_body').html(htmlBody);
        }

        /**
         * Display the pheno information.
         * @param {Object} geneData - gene information
         */
        function displayPhenoData(phenoID) {
            console.log('Displaying pheno information:', phenoID);

            let pheno = g.dataSets[g.dataSetID].phenotypes[phenoID];

            let htmlHeader = phenoID;
            let htmlBody = pheno.desc;
            $('#div_current_item_information_header').html(htmlHeader);
            $('#div_current_item_information_body').html(htmlBody);
        }

        function updatePeaks(minValue) {
            g.LOD = minValue;
            let newData = [];

            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                $.each(g.lodPeaks, function (index, element) {
                    let markerChr = g.chromosomes.chr[element[1]];
                    let markerPos = element[2] * 1000000.0;
                    let geneChr = g.chromosomes.chr[element[5]];
                    let genePos = element[6] * 1000000.0;

                    if ((markerChr !== undefined) && (geneChr !== undefined)) {
                        //data.push([+(geneChr.start + genePos), +(markerChr.start + markerPos)]);

                        if (element[7] >= g.LOD) {
                            newData.push({
                                geneChr: geneChr,
                                genePos: genePos,
                                genePosOrig: element[2],
                                markerChr: markerChr,
                                markerPos: markerPos,
                                x: +(markerChr.start + markerPos),
                                y: +(geneChr.start + genePos),
                                lod: element[7],
                                markerID: element[0],
                                geneID: element[3],
                                geneSymbol: element[4]
                            });
                        }
                    }
                });
            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                // [marker_id, chrom, position, protein_id, gene_id, symbol, gene_chrom, gene_mid, lod
                // ["1_106129060","1",106.12906,"ENSMUSP00000108356","ENSMUSG00000009907","Vps4b","1",106.7804,8.2928352859],
                $.each(g.lodPeaks, function (index, element) {
                    let markerChr = g.chromosomes.chr[element[1]];
                    let markerPos = element[2] * 1000000.0;
                    let geneChr = g.chromosomes.chr[element[6]];
                    let genePos = element[7] * 1000000.0;

                    if ((markerChr !== undefined) && (geneChr !== undefined)) {
                        //data.push([+(geneChr.start + genePos), +(markerChr.start + markerPos)]);

                        if (element[8] >= g.LOD) {
                            newData.push({
                                geneChr: geneChr,
                                genePos: genePos,
                                genePosOrig: element[2],
                                markerChr: markerChr,
                                markerPos: markerPos,
                                x: +(markerChr.start + markerPos),
                                y: +(geneChr.start + genePos),
                                lod: element[8],
                                markerID: element[0],
                                proteinID: element[3],
                                geneID: element[4],
                                geneSymbol: element[5]
                            });
                        }
                    }
                });
            } else {
                //error
            }

            //console.log('g.chartPeaks=', g.chartPeaks);
            //console.log('g.chartPeaks.series[0].=', g.chartPeaks.series[0]);
            // TODO: https://github.com/highcharts/highcharts/issues/8355
            g.chartPeaks.series[0].setData(newData, true, false, false);
        }

        function exportLODPeaks(id, peaks) {
            let csvContent = '';

            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                csvContent = '"gene_id","symbol","gene_chrom","gene_mid","marker_id","marker_chrom","marker_position","lod"\n';
                // [marker_id, chrom, position, gene_id, symbol, gene_chrom, gene_mid, lod
                //     0         1       2        3       4         5          6        7
                // ["1_100007442","1",100.0074,"ENSMUSG00000028028","Alpk1","3",127.7255,6.1147]
                $.each(peaks, function(k, v) {
                    let line = `"${v[3]}","${v[4]}","${v[5]}",${v[6]},"${v[0]}","${v[1]}",${v[2]},"${v[7]}"\n`;
                    csvContent += line;
                });
            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                csvContent = '"protein_id","gene_id","symbol","gene_chrom","gene_mid","marker_id","marker_chrom","marker_position","lod"\n';
                // [marker_id, chrom, position, protein_id, gene_id, symbol, gene_chrom, gene_mid, lod
                //     0         1       2          3          4        5         6        7        8
                // ["1_106129060","1",106.12906,"ENSMUSP00000108356","ENSMUSG00000009907","Vps4b","1",106.7804,8.2928352859],
                $.each(peaks, function(k, v) {
                    let line = `"${v[3]}","${v[4]}","${v[5]}","${v[6]}",${v[7]},"${v[0]}","${v[1]}",${v[2]},"${v[8]}"\n`;
                    csvContent += line;
                });
            } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                csvContent = '"marker_id","marker_chrom","marker_position","phenotype","lod"\n';
                // ["5_137006393","5",137.0064,"MEcyan","MEcyan","MEcyan",9.1161],[
                //       0         1     2        3        4        5        6
                $.each(peaks, function(k, v) {
                    let line = `"${v[0]}","${v[1]}",${v[2]},"${v[4]}",${v[6]}\n`;
                    csvContent += line;
                });
            }

            downloadCSV(csvContent, `${id}_LODPEAKS.csv`, 'text/csv;encoding:utf-8');
        }

        /**
         * Plot LOD Peaks as a scatter plot.
         */
        function plotLODPeaks() {
            // DATA will depend on type
            let ticksStartVals = [];
            let ticksStartText = {};
            let ticksMidVals = [];
            let ticksMidText = {};

            let firstDraw = true;

            let plotLines = [];
            let plotBands = [];
            let i = 0;

            $.each(g.chromosomes.chr, function(idx, value) {
                plotLines.push({
                   color: '#eeeeee',
                   value: value.start,
                   width: 2,
                });

                if ((i % 2) === 1) {
                    plotBands.push({
                        color: '#eeeeee',
                        from: value.start,
                        to: value.start + value.length
                    });
                }

                i += 1;

                ticksStartVals.push(value.start);
                ticksStartText[value.start] = value.chromosome;
                ticksMidVals.push(value.mid);
                ticksMidText[value.mid] = value.chromosome;
            });

            if ((i % 2) === 0) {
                //console.log('adding ', g.chromosomes.idx[i-1].chromosome);
                plotBands.push({
                    color: '#eeeeee',
                    from: g.chromosomes.idx[i-1].start,
                    to: g.chromosomes.idx[i-1].start + g.chromosomes.idx[i-1].length
                });
            }

            console.log('plotBands=', plotBands);

            plotLines.push({
               color: '#eeeeee',
               value: g.chromosomes.totalLength,
               width: 2,
            });

            //console.log('ticksStartVals=', ticksStartVals);
            //console.log('ticksStartText=', ticksStartText);
            //console.log('ticksMidVals=', ticksMidVals);
            //console.log('ticksMidText=', ticksMidText);

            let data = [];
            let xAxis = null;
            let yAxis = null;
            let tooltip = null;
            let clickFunction = null;
            let series = [];
            let legend = null;
            let minLOD = Infinity;
            let maxLOD = -Infinity;
            let html = `<div class="row">
                            <div class="col">
                                <div id="plotLodPeaks" style="width: 100%"></div>
                            </div>
                        </div>`;

            g.LOD = null;

            let exporting = appendExportButton('Download CSV', function() {
                exportLODPeaks(g.dataSets[g.dataSetID].displayName, g.lodPeaks);
            });
            exporting.filename = g.dataSets[g.dataSetID].displayName + '_PEAKS';

            //TODO: make sure g.LOD is defined above and g.chartPeaks and reset appropriately

            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                // [marker_id, chrom, position, gene_id, symbol, gene_chrom, gene_mid, lod
                // ["1_100007442","1",100.0074,"ENSMUSG00000028028","Alpk1","3",127.7255,6.1147]
                $.each(g.lodPeaks, function(index, element) {

                    let markerChr = g.chromosomes.chr[element[1]];
                    let markerPos = element[2] * 1000000.0;
                    let geneChr = g.chromosomes.chr[element[5]];
                    let genePos = element[6] * 1000000.0;

                    if ((markerChr !== undefined) && (geneChr !== undefined)) {
                        //data.push([+(geneChr.start + genePos), +(markerChr.start + markerPos)]);

                            data.push({
                                geneChr: geneChr,
                                genePos: genePos,
                                genePosOrig: element[6],
                                markerChr: markerChr,
                                markerPos: markerPos,
                                markerPosOrig: element[2],
                                x: +(markerChr.start + markerPos),
                                y: +(geneChr.start + genePos),
                                lod: element[7],
                                markerID: element[0],
                                geneID: element[3],
                                geneSymbol: element[4]
                            });


                        minLOD = Math.min(minLOD, element[7]);
                        maxLOD = Math.max(maxLOD, element[7]);
                    }
                });

                minLOD = Math.max(0, Math.floor(minLOD));

                // maxLOD will be 80% of max lod score
                maxLOD = maxLOD * 0.8;

                // round up to nearest 10, 11->20, 7->10, 88->90
                maxLOD= Math.ceil(maxLOD / 10) * 10;

                let currentLOD = minLOD;


                console.log('minLOD=', minLOD, 'maxLOD=', maxLOD)

                html = html + `
                    <div class="form-group row">
                        <label for="lodSliderValue" class="col-sm-6 col-form-label">LOD Threshold</label>
                        <div class="col-sm">
                            <div class="input-group">
                                <input type="number" min="${minLOD}" max="${maxLOD}" step="1" class="form-control form-control-sm" id="lodSliderValue" placeholder="">
                                <span class="input-group-btn">
                                    <button id="btnLODSet" class="btn btn-primary btn-sm">Set</i></button>
                                </span>
                            </div>
                        </div>
                    </div>
                    <div class="row">
                        <div class="col-2 text-right">
                            <span>${minLOD}</span>
                        </div>
                        <div class="col-8">
                            <div class="nstSlider w-100"
                                 data-range_min="${minLOD}"
                                 data-range_max="${maxLOD}"
                                 data-cur_min="${currentLOD}">
                                 <div class="leftGrip"></div>
                            </div>
                        </div>
                        <div class="col-2 text-left">
                            <span>${maxLOD}</span>
                        </div>
                    </div>`;

                $('#tabLodPeaks').html(html);

                let nstSlider = $('.nstSlider').nstSlider({
                    left_grip_selector: '.leftGrip',
                    value_changed_callback: function(cause, leftValue, rightValue) {
                        $('#lodSliderValue').val(leftValue);
                    },
                    user_mouseup_callback: function(minValue, maxValue, isLeftGrip, didValuesChange) {
                        //console.log(minValue, maxValue, isLeftGrip, didValuesChange);
                        updatePeaks(minValue);
                    }
                });

                $('#btnLODSet').on('click', function(evt) {
                    evt.preventDefault();
                    let sliderVal = $('#lodSliderValue').val().trim();

                    if (sliderVal) {
                        sliderVal = +sliderVal;
                        if (sliderVal < minLOD) {
                            sliderVal = minLOD;
                        } else if (sliderVal > maxLOD) {
                            sliderVal = maxLOD;
                        }
                        $('.nstSlider').nstSlider('set_position', sliderVal);
                        updatePeaks(sliderVal);
                    }
                });

                $('#lodSliderValue').keypress(function(evt) {
                    let code = evt.which ? evt.which : evt.keyCode;
                    if (code === 13) {
                        $(this).blur();
                        $('#btnLODSet').focus().click();
                    }
                });

                series = [{
                    data: data,
                    // When a series contains a data array that is
                    // longer than this, only one dimensional arrays
                    // of numbers, or two dimensional arrays with
                    // x and y values are allowed.
                    // Also, only the first point is tested, and
                    // the rest are assumed to be the same format.
                    // This saves expensive data checking and
                    // indexing in long series. Set it to 0 disable.
                    // Defaults to 1000.
                    turboThreshold: 0,
                    //boostThreshold: 0,
                    //fillOpacity: 0.1,
                    //color: 'rgba(223, 83, 83, 0.5)',
                }];

                legend = {
                    enabled: false
                };

                xAxis = {
                    endOnTick: false,
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return ticksMidText[this.value];
                        }
                    },
                    min: 0,
                    max: g.chromosomes.totalLength,
                    plotLines: plotLines,
                    showLastLabel: true,
                    startOnTick: false,
                    tickPositions: ticksMidVals,
                    title: {
                        enabled: true,
                        text: 'Marker'
                    },
                };

                yAxis = {
                    endOnTick: false,
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return ticksMidText[this.value];
                        }
                    },
                    min: 0,
                    max: g.chromosomes.totalLength,
                    plotLines: plotLines,
                    showLastLabel: true,
                    startOnTick: false,
                    tickPositions: ticksMidVals,
                    title: {
                        enabled: true,
                        text: 'Gene'
                    },
                };

                clickFunction = function (event) {
                    selectGeneProtein(this.geneID, null, g.dataSetID);
                };

                tooltip = {
                    followPointer: false,
                    formatter: function (a) {
                        return `LOD: ${this.point.lod}<br>Gene ID: ${this.point.geneID}<br>Symbol: ${this.point.geneSymbol}<br>Gene Location: ${this.point.geneChr.chromosome}:${this.point.genePos}<br>Marker ID: ${this.point.markerID}<br>Marker Location: ${this.point.markerChr.chromosome}:${this.point.markerPos}`;
                    }
                };



            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                // [marker_id, chrom, position, protein_id, gene_id, symbol, gene_chrom, gene_mid, lod
                // ["1_106129060","1",106.12906,"ENSMUSP00000108356","ENSMUSG00000009907","Vps4b","1",106.7804,8.2928352859],
                $.each(g.lodPeaks, function(index, element) {

                    let markerChr = g.chromosomes.chr[element[1]];
                    let markerPos = element[2] * 1000000.0;
                    let geneChr = g.chromosomes.chr[element[6]];
                    let genePos = element[7] * 1000000.0;

                    if ((markerChr !== undefined) && (geneChr !== undefined)) {
                        //data.push([+(geneChr.start + genePos), +(markerChr.start + markerPos)]);

                        data.push({
                            geneChr: geneChr,
                            genePos: genePos,
                            genePosOrig: element[7],
                            markerChr: markerChr,
                            markerPos: markerPos,
                            markerPosOrig: element[2],
                            x: +(markerChr.start + markerPos),
                            y: +(geneChr.start + genePos),
                            lod: element[8],
                            markerID: element[0],
                            proteinID: element[3],
                            geneID: element[4],
                            geneSymbol: element[5]
                        });

                        minLOD = Math.min(minLOD, element[8]);
                        maxLOD = Math.max(maxLOD, element[8]);
                    }
                });

                minLOD = Math.max(0, Math.floor(minLOD));

                // maxLOD will be 80% of max lod score
                maxLOD = maxLOD * 0.8;

                // round up to nearest 10, 11->20, 7->10, 88->90
                maxLOD = Math.ceil(maxLOD / 10) * 10;

                let currentLOD = minLOD;

                html = html + `
                    <div class="form-group row">
                        <label for="lodSliderValue" class="col-sm-6 col-form-label">LOD Threshold</label>
                        <div class="col-sm">
                            <div class="input-group">
                                <input type="number" min="${minLOD}" max="${maxLOD}" step="1" class="form-control form-control-sm" id="lodSliderValue" placeholder="">
                                <span class="input-group-btn">
                                    <button id="btnLODSet" class="btn btn-primary btn-sm">Set</i></button>
                                </span>
                            </div>
                        </div>
                    </div>
                    <div class="row">
                        <div class="col-2 text-right">
                            <span>${minLOD}</span>
                        </div>
                        <div class="col-8">
                            <div class="nstSlider w-100"
                                 data-range_min="${minLOD}"
                                 data-range_max="${maxLOD}"
                                 data-cur_min="${currentLOD}">
                                 <div class="leftGrip"></div>
                            </div>
                        </div>
                        <div class="col-2 text-left">
                            <span>${maxLOD}</span>
                        </div>
                    </div>`;

                $('#tabLodPeaks').html(html);

                let nstSlider = $('.nstSlider').nstSlider({
                    left_grip_selector: '.leftGrip',
                    value_changed_callback: function(cause, leftValue, rightValue) {
                        $('#lodSliderValue').val(leftValue);
                    },
                    user_mouseup_callback: function(minValue, maxValue, isLeftGrip, didValuesChange) {
                        //console.log(minValue, maxValue, isLeftGrip, didValuesChange);
                        updatePeaks(minValue);
                    }
                });

                $('#btnLODSet').on('click', function(evt) {
                    evt.preventDefault();
                    let sliderVal = $('#lodSliderValue').val().trim();

                    if (sliderVal) {
                        sliderVal = +sliderVal;
                        if (sliderVal < minLOD) {
                            sliderVal = minLOD;
                        } else if (sliderVal > maxLOD) {
                            sliderVal = maxLOD;
                        }
                        $('.nstSlider').nstSlider('set_position', sliderVal);
                        updatePeaks(sliderVal);
                    }
                });

                $('#lodSliderValue').keypress(function(evt) {
                    let code = evt.which ? evt.which : evt.keyCode;
                    if (code === 13) {
                        $(this).blur();
                        $('#btnLODSet').focus().click();
                    }
                });

                series = [{
                    // When a series contains a data array that is
                    // longer than this, only one dimensional arrays
                    // of numbers, or two dimensional arrays with
                    // x and y values are allowed.
                    // Also, only the first point is tested, and
                    // the rest are assumed to be the same format.
                    // This saves expensive data checking and
                    // indexing in long series. Set it to 0 disable.
                    // Defaults to 1000.
                    turboThreshold: 0,
                    //boostThreshold: 0,
                    //fillOpacity: 0.1,
                    //color: 'rgba(223, 83, 83, 0.5)',
                    data: data
                }];

                legend = {
                    enabled: false
                };

                xAxis = {
                    endOnTick: false,
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return ticksMidText[this.value];
                        }
                    },
                    min: 0,
                    max: g.chromosomes.totalLength,
                    plotLines: plotLines,
                    showLastLabel: true,
                    startOnTick: false,
                    tickPositions: ticksMidVals,
                    title: {
                        enabled: true,
                        text: 'Marker'
                    },
                };

                yAxis = {
                    endOnTick: false,
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return ticksMidText[this.value];
                        }
                    },
                    min: 0,
                    max: g.chromosomes.totalLength,
                    plotLines: plotLines,
                    showLastLabel: true,
                    startOnTick: false,
                    tickPositions: ticksMidVals,
                    title: {
                        enabled: true,
                        text: 'Gene'
                    },
                };


                clickFunction = function (event) {
                    selectGeneProtein(this.geneID, this.proteinID, g.dataSetID);
                };

                tooltip = {
                    followPointer: false,
                    formatter: function (a) {
                        return `LOD: ${this.point.lod}<br>Protein ID: ${this.point.proteinID}<br>Gene ID: ${this.point.geneID}<br>Symbol: ${this.point.geneSymbol}<br>Gene Location: ${this.point.geneChr.chromosome}:${this.point.genePos}<br>Marker ID: ${this.point.markerID}<br>Marker Location: ${this.point.markerChr.chromosome}:${this.point.markerPos}`;
                    }
                };



            } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                // ["5_137006393","5",137.0064,"MEcyan","MEcyan","MEcyan",9.1161],[

                let seriesData = {};
                //console.log('g.lodPeaks=', g.lodPeaks);
                $.each(g.lodPeaks, function(index, element) {
                    // element[0] = marker_id
                    let markerChr = g.chromosomes.chr[element[1]];
                    let markerPos = element[2] * 1000000.0;

                    // lookup the phenotype
                    let phenotype = g.dataSets[g.dataSetID].phenotypes[element[3]];

                    if (!(phenotype.category in seriesData)) {
                        seriesData[phenotype.category] = [];
                    }

                    seriesData[phenotype.category].push({
                        x: markerChr.start + markerPos,
                        y: element[6],
                        lod: element[6],
                        phenoDataName: element[3],
                        phenoShortName: element[4],
                        phenoDescription: element[5],
                        markerPosOrig: element[2],
                        markerChr: markerChr,
                        markerPos: markerPos,
                        markerID: element[0]
                    });
                    minLOD = Math.min(minLOD, element[6]);
                    maxLOD = Math.max(maxLOD, element[6]);
                });

                console.log('minLOD=', minLOD, 'maxLOD=', maxLOD);

                minLOD = Math.floor(minLOD);
                maxLOD = Math.ceil(maxLOD);

                console.log('ADJUSTED: minLOD=', minLOD, 'maxLOD=', maxLOD);

                $('#tabLodPeaks').html(html);

                console.log('seriesData=', seriesData);

                series = [];
                $.each(seriesData, function(index, element) {
                    //console.log(index, element);
                    series.push({
                        data: element,
                        name: index,
                        turboThreshold: 0,
                    })
                });

                legend = {
                    align: 'center',
                    floating: false,
                    layout: 'horizontal',
                    verticalAlign: 'bottom',
                };

                xAxis = {
                    endOnTick: false,
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return ticksMidText[this.value];
                        }
                    },
                    min: 0,
                    max: g.chromosomes.totalLength,
                    plotBands: plotBands,
                    plotLines: plotLines,
                    showLastLabel: true,
                    startOnTick: false,
                    tickPositions: ticksMidVals,
                    title: {
                        enabled: true,
                        text: 'Marker Location'
                    },
                };

                yAxis = {
                    title: {
                        enabled: true,
                        text: 'LOD'
                    },
                    min: minLOD,
                    max: maxLOD,
                    tickWidth: 0,
                    startOnTick: true,
                    endOnTick: true,
                    showLastLabel: true
                };


                console.log('yAxis=', yAxis);

                clickFunction = function (event) {
                    selectPhenotype(this.phenoDataName, g.dataSetID);
                };

                tooltip = {
                    followPointer: false,
                    formatter: function () {
                        return `LOD: ${this.point.lod}<br>Phenotype: ${this.point.phenoShortName}<br>Category: ${this.series.name}<br><br>Marker ID: ${this.point.markerID}<br>Marker Location:${this.point.markerChr.chromosome}:${this.point.markerPos}`;
                    }
                };

            } else {
                // TODO: handle error
                console.error('WHAT TO DO????????????????????????????');
            }

            g.chartPeaks = Highcharts.chart('plotLodPeaks', {
                boost: {
                    debug: {
                        showSkipSummary: true
                    },
                    enabled: true,
                    useGPUTranslations: true,
                },
                chart: {
                    type: 'scatter',
                    zoomType: 'xy',
                },
                exporting: exporting,
                legend: legend,
                plotOptions: {
                    scatter: {
                        marker: {
                            radius: 2,
                            states: {
                                hover: {
                                    enabled: true,
                                    lineColor: 'rgb(100,100,100)'
                                }
                            }
                        },
                        states: {
                            hover: {
                                marker: {
                                    enabled: false
                                }
                            }
                        }
                    },
                    series: {
                        point: {
                            events: {
                                click: clickFunction,
                                mouseOver: function(evt) {
                                    if(this.series.halo) {
                                        this.series.halo.attr({
                                           'class': 'highcharts-tracker'
                                        }).toFront();
                                    }
                                }
                            }
                        }
                    }
                },
                series: series,
                subtitle: {
                    text: ''
                },
                title: {
                    text: 'LOD Peaks'
                },
                tooltip: tooltip,
                xAxis: xAxis,
                yAxis: yAxis,
            });

            if ((g.dataSets[g.dataSetID].dataType === 'mRNA') ||
                (g.dataSets[g.dataSetID].dataType === 'protein')) {
                // this hack is awful and needs to be fixed
                // this is the only way I could get the tooltips to appear
                // without manually moving the slider
                // I'm guessing it has something to do with canvas vs svg
                setTimeout(function () {
                    console.log('updatingPeaks 1');
                    updatePeaks(maxLOD);
                    setTimeout(function () {
                        console.log('updatingPeaks 2');
                        updatePeaks(minLOD);
                    }, 250);
                }, 250);
            }
        }

        function startTask() {
            g.runningTask = true;
            swal({
                allowEnterKey: false,
                allowEscapeKey: false,
                allowOutsideClick: false,
                buttonsStyling: false,
                cancelButtonClass: 'btn btn-danger',
                html: $('#ap').html(),
                showConfirmButton: false,
                showCancelButton: false,
                showCloseButton: true,
                title: 'Processing...',
                width: '20rem',
                onClose: function() {
                      console.log('onClose', this);
                }
            }).then((result) => {
                console.log('then called');
                // result.dismiss can be 'cancel', 'overlay',
                // 'close', and 'timer'
                if (result.dismiss === 'close') {
                    stopTask();
                }
            });
        }

        function stopTask() {
            g.runningTask = false;
            swal.close();
        }

        function downloadChromosomes(version, species, callback) {
            let chromosomeURL = `{{ url_for('api.api_get', url='', _external=True) }}http://churchill-lab.jax.org/ensimpl/api/chromosomes?version=${version}`;
            if (species != null) {
                chromosomeURL += '&species=' + species;
            }
            downloadData(chromosomeURL, 'chromosomes', callback);
        }

        function downloadLODPeaks(dataSetID, callback) {
            let lodPeaksURL = `{{ url_for('api.api_get', url=config.API_R_BASE + '/lodpeaks', _external=True) }}?dataset=${dataSetID}`;
            downloadData(lodPeaksURL, 'LODPeaks', callback);
        }

        function downloadDataSets(callback) {
            let dataSetsURL = '{{ url_for('api.api_get', url=config.API_R_BASE + '/datasets', _external=True) }}';
            downloadData(dataSetsURL, 'dataSets', callback);
        }

        function getTextEnd(text, chart, s) {
            let G_ELEM = document.getElementById('getTheFontSize');
            let J_ELEM = $('#getTheFontSize');
            J_ELEM.text(text);
            let computedLength = G_ELEM.getComputedTextLength();
            let temp = chart.xAxis[0].toPixels(s);
            let ret = chart.xAxis[0].toValue(temp + computedLength + 10);
            //console.log('getTextEnd=', text, computedLength);
            //console.log('getTextEnd=', temp, ret);
            return ret;
        }


        function fitSNPGenes(chart, geneData) {
            let genes = [];
            $.each(geneData.matches, function(index, value) {
                if (value.symbol !== null) {
                    value.text_end = getTextEnd(value.symbol, chart, value.position_start);
                    value.max_end = Math.max(value.position_end, value.text_end);
                    value.id = value.ensembl_id;

                    //console.log('Loop=', value.symbol, value.position_start, value.position_end, value.text_end, value.max_end);
                    genes.push(value);
                }
            });

            // generate the gene layout
            genes.sort(function(a, b) {
                return a.position_start - b.position_start;
            });


            layoutFeatures(genes);

            genes.sort(function(a, b) {
                return a.position_start - b.position_start;
            });

            let categories = [];
            let data = [];

            $.each(genes, function(index, value) {
                if (!(value.slot in categories)) {
                    categories.push(value.slot * -1);
                }
                let e = value;
                e.y = value.slot * -1;
                e.x = value.position_start;
                e.x2 = value.position_end;
                data.push(e);
                //console.log(e);
            });

            chart.series[1].setData(data);
        }

        function exportSNPAssocData(id, snps, genes) {
            // TODO: export genes?
            let csvContent = `"id","position","alleles","csq","lod"\n`;

            $.each(snps, function(k, v) {
                let pos = v.x / 1000000.0;
                let line = `"${v['id']}",${pos},"${v['alleles']}","${v['csq']}",${v['lod']}\n`;
                csvContent += line;
            });

            downloadCSV(csvContent, `${id}_SNPASSOC.csv`, 'text/csv;encoding:utf-8');
        }

        /**
         * Plot the SNP Association plot.
         * @param {Array} snpData - array of SNP information
         * @param {Array} geneData - array of gene objects
         */
        function plotSNPAssoc(snpData, geneData, id, chromosome, location) {
            console.log('plotSNPAssoc');

            let dataByChrom = {};
            let maxLODScore = null;
            let xAxisTitle = '';

            let plotTitle = '';
            let currentID = '';

            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                plotTitle = `${g.geneID} (${g.currentGene.gene[g.geneID].symbol})`;
                currentID = g.geneID;
            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                plotTitle = `${g.proteinID} (${g.currentGene.gene[g.geneID].symbol})`;
                currentID = g.proteinID;
            } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                plotTitle = g.phenotypeID;
                currentID = g.phenotypeID;
            } else {
                // TODO: handle error
                console.error('MAJOR PROBLEM');
            }

            $('#snpWindowMenu').html(`
                <div class="col">
                    <button id="snpShiftLeft" class="btn btn-secondary"><i class="fas fa-arrow-left"></i> Shift Plot Left</button>
                </div>
                <div class="col">
                    <button id="snpShiftRight" class="btn btn-secondary float-right">Shift Plot Right <i class="fas fa-arrow-right"></i></button>
                </div>`);


            $('#snpShiftLeft').button().on('click', function(event) {
                generateSNPAssocPlot(id, chromosome, location - 1);
            });

            $('#snpShiftRight').button().on('click', function(event) {
                generateSNPAssocPlot(id, chromosome, location + 1);
            });

            snpData.sort(function(a, b) {
                return a.pos - b.pos;
            });

            let snps = [];

            if (snpData) {
                maxLODScore = snpData[0][10];
            }

            let minn = Infinity;
            let maxx = -Infinity;

            /*
         0   {"snp":"rs245710663",
         1   "chr":"1",
         2   "pos":14.5001,
         3   "alleles":"G|C",
         4   "sdp":64,
         5   "ensembl_gene":"",
         6   "csq":"intergenic_variant",
         7   "index":1,
         8   "interval":147,
         9   "on_map":false,
         10   "lod":0.0015},
*/

            $.each(snpData, function(index, element) {
                if (element[10] > maxLODScore) {
                    maxLODScore = element[10];
                }

                snps.push({
                    alleles: element[3],
                    csq: element[6],
                    id: element[0],
                    lod: element[10],
                    x: element[2] * 1000000,
                    y: 1.0 * element[10],
                });

                minn = Math.min(minn, element[2] * 1000000.0);
                maxx = Math.max(maxx, element[2] * 1000000.0);

            });

            maxLODScore += 1;

            let series = [];

            series.push({
                name: 'seriesSNPs',
                data: snps,
                turboThreshold: 0,
                color: 'rgba(223, 83, 83, .5)',
                yAxis: 'snpTop',
                type: 'scatter',
                marker: {
                    enabled: true,
                    radius: 2
                },

                /* Set the point threshold for when a series should enter boost mode.
                   Setting it to e.g. 2000 will cause the series to enter boost mode when there are 2000 or more points in the series.
                   To disable boosting on the series, set the boostThreshold to 0. Setting it to 1 will force boosting.
                   Requires modules/boost.js.
                   Defaults to 5000.
                */
                boostThreshold: 1,
                findNearestPointBy: 'xy',
            });

            series.push({
                name: 'seriesGenes',
                data: [],
                yAxis: 'genesBottom',
                type: 'xrange',
                dataLabels: {
                    align: 'left',
                    allowOverlap: true,
                    color: 'black',
                    defer: false,
                    enabled: true,
                    formatter: function() {
                        return this.point.symbol;
                    },
                    inside: true,
                    padding: 0,
                    style: {
                        textOutline: false
                    },
                },
                //boostThreshold: 0,
                /* When a series contains a data array that is longer than this,
                   only one dimensional arrays of numbers, or two dimensional
                   arrays with x and y values are allowed. Also, only the first
                   point is tested, and the rest are assumed to be the same
                   format. This saves expensive data checking and indexing in
                   long series. Set it to 0 disable.
                   Defaults to 1000.
                  */
                /*
                turboThreshold: 0,
                animation: {
                    duration: 0
                }
                */
            });

            let exporting = appendExportButton('Download CSV', function() {
                exportSNPAssocData(currentID, snps, geneData);
            });
            exporting.filename = currentID + '_SNPASSOC';

            g.chartSNPAssoc = Highcharts.chart({
                boost: {
                    debug: {
                        showSkipSummary: true
                    },
                    enabled: true,
                    useGPUTranslations: true,
                },
                chart: {
                    renderTo: 'plotSNPAssoc',
                    panning: true,
                    panKey: 'shift',
                    zoomType: 'x',
                    animation: false,
                    events: {
                        load: function () {
                            fitSNPGenes(this, geneData);
                        }
                    }
                },

                title: {
                    text: plotTitle,
                },

                xAxis: [{
                    min: minn,
                    max: maxx,
                    labels: {
                        enabled: true
                    },
                    title: {
                        text: ''
                    },
                    scrollbar: {
                        enabled: true
                    },
                    crosshair: {
                        snap: false
                    }
                }],

                yAxis: [{
                        id: 'snpTop',
                        title: {
                            text: 'LOD'
                        },
                        max: maxLODScore,
                        min: 0,
                        offset: 0,
                        height: '50%',
                    /*
                        labels: {
                            formatter: function () {
                                return this.value;
                            }
                        }
                        */
                    }, {
                        id: 'genesBottom',
                        reverse: true,
                        min: -12,
                        max: -0,
                        categories: [-0, -1, -2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12],
                        title: {
                            text: ""
                        },
                        visible: false,
                        offset: 0,
                        top: '51%',
                        height: '49%',
                    }
                ],

                plotOptions: {
                    scatter: {
                        marker: {
                            states: {
                                hover: {
                                    enabled: false,
                                    lineColor: 'rgb(100,100,100)'
                                }
                            }
                        },
                        states: {
                            hover: {
                                marker: {
                                    enabled: false
                                }
                            }
                        }
                    },

                    series: {
                        borderRadius: 0
                    },

                    xrange: {
                        borderWidth: 1,
                        borderColor: 'black',
                        colorByPoint: true,
                        colors: ['#7cb5ec'],
                        dataLabels: {
                            style: {
                                color: 'white'
                            }
                        }
                    }
                },

                legend: {
                    enabled: false
                },

                series: series,

                tooltip: {
                    positioner: function(labelWidth, labelHeight, point) {
                        let tooltipX, tooltipY;
                        let chart = g.chartSNPAssoc;

                        if (point.plotX + labelWidth > chart.plotWidth) {
                            tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                        } else {
                            tooltipX = point.plotX + chart.plotLeft + 20;
                        }

                        if (point.plotY + labelHeight > chart.plotHeight) {
                            tooltipY = point.plotY + chart.plotTop - labelHeight - 20;
                        } else {
                            tooltipY = point.plotY + chart.plotTop - 20;
                        }

                        return {
                            x: tooltipX,
                            y: tooltipY
                        };
                    },
                    useHTML: true,
                    shared: false,
                    followPointer: true,
                    enabled: true,
                    formatter: function () {
                        let text = "";
                        if (this.series.name === 'seriesSNPs') {
                            let csq = this.point.csq.split(',');
                            let csqs = csq.join('<br/>');

                            text = '<b>' + this.point.id + '</b><br/>' +
                                'LOD: ' + this.point.lod + '<br/>' +
                                'Position: ' + formatMbp(this.point.y / 100000) + '<br/>' +
                                'Alleles: ' + this.point.alleles + '<br/>' +
                                'CSQ: ' + csqs;
                        } else {
                            text = '<b>' + this.point.symbol + '</b><br/>' +
                                'Ensembl ID: ' + this.point.ensembl_gene_id + '<br/>' +
                                'Name: ' + this.point.name + '<br/>' +
                                'Position: ' + this.point.match_value + ' (' + this.point.strand + ')';

                        }
                        return text;
                    }
                },
                exporting: exporting,
            });
        }

        /**
         * Start downloading snp
         */
        function generateSNPAssocPlot(id, chromosome, location) {
            // TODO: windowSize needs to be figured out
            let windowSize = 1000000;
            let newLocation = Math.round(location * 1000000.0);
            let minPos = Math.max(newLocation - windowSize, 0);
            let maxPos = newLocation + windowSize;

            let snpAssocURL = `{{ config.API_R_BASE }}/snpassoc?dataset=${g.dataSetID}&id=${id}&chrom=${chromosome}&location=${newLocation}&windowSize=${windowSize}`;
            let genesURL = `http://churchill-lab.jax.org/ensimpl/api/search?species=${g.speciesID}&version=${g.ensemblVersion}&term=${chromosome}:${minPos}-${maxPos}`;

            {% if config.API_R_NUM_CORES %}
            snpAssocURL += '&nCores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'snpAssoc',
                    url: snpAssocURL
                }, {
                    url_id: 'genes',
                    url: genesURL
                }]};


            startTask();

            // reset the chart and clear it
            g.chartSNPAssoc = null;
            $('#plotSNPAssoc').html('');
            $('#snpWindowMenu').html('');


            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    console.log('data.group_id=', data.group_id);
                    updateSNPAssocData(data.group_id, id, chromosome, location);
                },
                error: function() {
                    // TODO: handle
                    alert('Unexpected error');
                }
            });
        }

        /**
         * Update the status of the downloading effect data.
         *
         * @param {string} groupID - the group identifier of the task
         */
        function updateSNPAssocData(groupID, id, chromosome, location) {
            // send GET request to status URL
            console.log('updateSNPAssocData');
            console.log(groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=True) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function (data, status, request) {
                        console.log('DATA=======', data);
                        if (data['status'] === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data['number_tasks_errors'] !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['error']}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data['number_tasks_errors'] === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['response']['error']}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the SNP Association Plot.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 2 datasets to get

                                    // 1. snpAssoc information

                                    console.log('show result');
                                    console.log('using data=', data);
                                    let info = data['response_data']['snpAssoc'];
                                    let len = info.response.result.length;

                                    let dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>SNP Points</strong></td><td>${len}</td></tr>
                                            </table>`;

                                    debugMessage('SNP Information', dbg);

                                    // 2. genes information
                                    info = data['response_data']['genes'];
                                    len = info.response.result.matches.length;

                                    dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Genes</strong></td><td>${len}</td></tr>
                                            </table>`;

                                    debugMessage('Genomic Information', dbg);

                                    plotSNPAssoc(data.response_data.snpAssoc.response.result, data.response_data.genes.response.result, id, chromosome, location);
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function () {
                                updateSNPAssocData(groupID, id, chromosome, location);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=True) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                    g.chartSNPAssoc = null;
                    $('#plotSNPAssoc').html('');
                    $('#snpWindowMenu').html('');
                });
            }
        }


        function exportEffectData(id, effectData, lodData) {
           console.log(effectData);
           console.log(lodData);

            let csvContent = `"id","chromosome","position"{%- for s in config.PLOT_EFFECT_STRAINS -%}
                ,"{{ s['name'] }}"
            {%- endfor -%},"lod"\n`;

            $.each(effectData, function(k, v) {
                let line = `"${v[0]}","${v[1]}",${v[2]},${v[3]},${v[4]},${v[5]},${v[6]},${v[7]},${v[8]},${v[9]},${v[10]},`;
                line += lodData[k].y;
                csvContent += `${line}\n`;
            });

            downloadCSV(csvContent, `${id}_EFFECT.csv`, 'text/csv;encoding:utf-8');
        }

        /**
         * Plot the effect data.
         * @param {Object} effectData - the effect data
         */
        function plotEffectData(effectData) {
            console.log('Plotting effect data');

            let strainInfo = {};
            {% for s in config.PLOT_EFFECT_STRAINS %}
                strainInfo['{{ s['key'] }}'] = {
                    'color':'{{ s['color'] }}',
                    'name':'{{ s['name'] }}',
                    'data':[]
                };
            {% endfor %}

            let chromosome = null;
            let allEffectValues = [];

            $.each(effectData, function(index, element) {
                //{"A":-0.1224,..."H":1.45,"chr":"4","id":"4_156339889","pos":156.3399}
                //id,chrom,pos,A-H
                chromosome = element[1];
                let position = element[2] * 1000000;

            {% for s in config.PLOT_EFFECT_STRAINS %}
                {# when data is expanded #}
                {# strainInfo['{{ s['key'] }}']['data'].push([position, element.{{ s['key'] }}]); #}
                {# allEffectValues.push(element.{{ s['key'] }}); #}

                {# when data is collapsed #}
                strainInfo['{{ s['key'] }}']['data'].push([position, element[{{ loop.index0 + 3}}]]);
                allEffectValues.push(element[{{ loop.index0 + 3}}]);
            {% endfor %}
            });

            let effectMinValue = Math.floor(Math.min.apply(null, allEffectValues));
            let effectMaxValue = Math.ceil(Math.max.apply(null, allEffectValues));

            /*
            if ((effectMaxValue > 30.0) || (effectMinValue < -30.0)) {
                displayError('Effect Plot Error', 'Effect values appear to be invalid, ' + effectMinValue + ' to ' + effectMaxValue);
                return;
            }
            */

            let xAxisTickNum = Math.floor(g.chromosomes.chr[chromosome].length / 20000000);
            let xAxisTickVals = [];
            let xAxisTickText = [];
            let series = [];

            for (let i = 1; i <= xAxisTickNum; i++) {
                xAxisTickVals.push(20000000 * i);
                xAxisTickVals.push((20 * i) + 'Mb')
            }

            $.each(strainInfo, function(key, value) {
                let s = {
                    color: value['color'],
                    data: value['data'],
                    lineWidth: 1,
                    marker: {
                        symbol: 'circle',
                    },
                    name: value['name'],
                    showInLegend: true,
                    stack: 0,
                    tooltip: {
                        shared: false,
                        headerFormat:'<b>{series.name}</b>',
                        pointFormatter: function() {
                            return '<br>Effect: ' + this.y + '<br>Position: ' +  (this.x/1000000);
                        }
                    },
                    turboThreshold: 0,
                    type: 'line',
                    yAxis:'axisEffect',
                };
                series.push(s);
            });

            let plotTitle = '';
            let currentID = '';

            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                plotTitle = `${g.geneID} (${g.currentGene.gene[g.geneID].symbol})`;
                currentID = g.geneID;
            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                plotTitle = `${g.proteinID} (${g.currentGene.gene[g.geneID].symbol})`;
                currentID = g.proteinID;
            } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                plotTitle = g.phenotypeID;
                currentID = g.phenotypeID;
            } else {
                // TODO: handle error
                console.error('MAJOR PROBLEM');
            }

            let lodData = [];
            let lodMinValue = Infinity;
            let lodMaxValue = -Infinity;

            // get the data from the LOD Chart to use here
            $.each(g.chartLOD.userOptions.series[0].data, function(index, element) {
                if (element.chr === chromosome) {
                    element.x = element.chrPos * 1000000.0;
                    lodData.push(element);

                    lodMinValue = Math.min(element.y, lodMinValue);
                    lodMaxValue = Math.max(element.y, lodMaxValue);
                }
            });

            console.log('lodData=', lodData);
            console.log('lodMinValue=', lodMinValue);
            console.log('lodMaxValue=', lodMaxValue);

            series.push({
                animation: false,
                color: '#000000',
                data: lodData,
                lineWidth: 1,
                name: 'LOD',
                showInLegend: false,
                stack: 0,
                tooltip: {
                    shared: false,
                    headerFormat: '',
                    pointFormatter: function() {
                        return '<br>LOD: ' + this.y + '<br>Position: ' +  (this.x/1000000);
                    }
                },
                turboThreshold: 1000000, // TODO: what should this value be
                yAxis:'axisLOD',
            });

            let lodTickVals = null;
            // round up to nearest 10, 11->20, 7->10, 88->90
            let lodMaxTick = Math.ceil(lodMaxValue / 10) * 10;

            // determine ticks if > 10, otherwise let the chart do it
            if (lodMaxValue > 10) {
                let lodMaxTick = lodMaxValue + 1;
                let maxNumTicks = Math.min(4, Math.ceil(lodMaxTick / 10) * 10);
                lodTickVals = [];

                let step = Math.ceil((lodMaxValue/maxNumTicks) / 10) * 10;
                for (let i = 0; i < lodMaxValue; i += step) {
                    lodTickVals.push(i);
                }
                lodTickVals.push(lodMaxTick);
            }

            let exporting = appendExportButton('Download CSV', function() {
                exportEffectData(currentID, effectData, lodData);
            });
            exporting.filename = currentID + '_EFFECT';

            g.chartEffect = Highcharts.chart({
                boost: {
                    useGPUTranslations: true
                },
                chart: {
                    renderTo: 'plotEffect',
                    zoomType: 'x'
                },
                exporting: exporting,
                legend: {
                    align: 'right',
                    backgroundColor: (Highcharts.theme && Highcharts.theme.legendBackgroundColor) || '#FFFFFF',
                    floating: false,
                    layout: 'vertical',
                    verticalAlign: 'top',
                    y: 100,
                },
                responsive: {
                    rules: [{
                        condition: {
                            maxWidth: 400
                        },
                        chartOptions: {
                            legend: {
                                align: 'center',
                                layout: 'horizontal',
                                verticalAlign: 'top',
                                y: 0,
                            },
                        }
                    }]
                },
                series: series,
                subtitle: {
                    text: 'Chromosome ' + chromosome,
                    verticalAlign: 'bottom',
                },
                title: {
                    text: plotTitle,
                },
                xAxis: {
                    crosshair: true,
                    gridLineWidth: 1,
                    lineWidth: 0.5,
                },
                yAxis: [{
                    height: '70%',
                    id: 'axisEffect',
                    labels: {
                        formatter: function() {
                            if (this.isFirst) {
                                return '';
                            }
                            return this.value;
                        }
                    },
                    maxPadding: 0,
                    offset: 0,
                    title: {
                        text: 'EFFECT'
                    },
                }, {
                    height: '30%',
                    id: 'axisLOD',
                    labels: {
                        formatter: function() {
                            if (this.isLast) {
                                return '';
                            }
                            return this.value;
                        }
                    },
                    maxPadding: 0,
                    offset: 0,
                    plotLines: [{
                        value: lodMaxTick,
                        color: 'black',
                        zIndex: 5,
                        width: 2
                    }],
                    tickPositions: lodTickVals,
                    title: {
                        text: 'LOD'
                    },
                    top: '70%',
                }],
            });
        }

        function exportMediationData(id, data) {
            let csvContent = `"gene_id","symbol","chromosome","position","lod"\n`;
            let protein = false;
            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                protein = false;
            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                protein = true;
                csvContent = '"protein_id",' + csvContent;
            }

            $.each(data, function(k, v) {
                let line = `"${v['gene_id']}","${v['symbol']}","${v['chr']}",${v['pos']},${v['LOD']}`;
                if (protein) {
                    line = `"${v['protein_id']}",` + line;
                }
                csvContent += `${line}\n`;
            });

            downloadCSV(csvContent, `${id}_MEDIATION.csv`, 'text/csv;encoding:utf-8');
        }


        /**
         * Plot mediation data
         * @param {Object} mediationData - mediation data
         */
        function plotMediation(mediationData, dsMediateAgainst) {
            let newData = [];

            if (g.dataSets[dsMediateAgainst].dataType === 'mRNA') {
                // ["ENSMUSG00000000001", "Gnai3", "3", 108.1267, 140.1666]
                $.each(mediationData, function(index, element) {
                    if (element[2] in g.chromosomes.chr) {
                        let chr = g.chromosomes.chr[element[2]];
                        element.x = chr.start + (element[3] * 1000000.0);
                        element.y = element[4];

                        newData.push({
                            chr: element[2],
                            gene_id: element[0],
                            LOD: element[4],
                            pos: element[3],
                            symbol: element[1],
                            x: chr.start + (element[3] * 1000000.0),
                            y: element[4]
                        });
                    } else {
                        //console.log(element);
                    }
                });

            } else if (g.dataSets[dsMediateAgainst].dataType === 'protein') {
                // ["ENSMUSP00000000001", "ENSMUSG00000000001", "Gnai3", "3", 108.126713, 7.1102046955]
                $.each(mediationData, function(index, element) {
                    if (element[3] in g.chromosomes.chr) {
                        let chr = g.chromosomes.chr[element[3]];
                        element.x = chr.start + (element[4] * 1000000.0);
                        element.y = element[5];

                        newData.push({
                            chr: element[3],
                            gene_id: element[1],
                            protein_id: element[0],
                            LOD: element[5],
                            pos: element[4],
                            symbol: element[2],
                            x: chr.start + (element[4] * 1000000.0),
                            y: element[5]
                        });
                    } else {
                        //console.log(element);
                    }
                });
            } else {
                // TODO: handle error
                console.error('MAJOR PROBLEM');
            }

            let axisTickVals = [];
            let axisTickText = [];

            // rectangles to display
            let dataChromosome = {};
            let chromosomeAxisVals = [];
            let chromosomeAxisText = {};
            let chromosomeBands = [];

            $.each(g.chromosomes.idx, function(index, element) {

                axisTickVals.push(element.mid);
                axisTickText.push(element.chromosome);

                dataChromosome[element.chromosome] = {};
                dataChromosome[element.chromosome].data = [];

                chromosomeAxisVals.push(element.mid);
                chromosomeAxisText[element.mid] = element.chromosome;

                // TODO: style this
                let fillColor = '#eeeeee';
                if ((index % 2) === 1) {
                    chromosomeBands.push({
                        color: fillColor,
                        from: element.start,
                        to: element.end
                    });
                }
            });

            let plotTitle = '';
            let currentID = '';

            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                plotTitle = `${g.geneID} (${g.currentGene.gene[g.geneID].symbol})`;
                currentID = g.geneID;
            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                plotTitle = `${g.proteinID} (${g.currentGene.gene[g.geneID].symbol})`;
                currentID = g.proteinID;
            } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                plotTitle = g.phenotypeID;
                currentID = g.phenotypeID;
            } else {
                // TODO: handle error
                console.error('MAJOR PROBLEM');
            }

            let exporting = appendExportButton('Download CSV', function() {
                exportMediationData(currentID, newData);
            });
            exporting.filename = currentID + '_MEDIATION';

            g.chartMediation = Highcharts.chart({
                chart: {
                    renderTo: 'plotMediation',
                    zoomType: 'x',
                },
                exporting: exporting,
                plotOptions: {
                    scatter: {
                        marker: {
                            radius: 5,
                            states: {
                                hover: {
                                    enabled: true,
                                    lineColor: 'rgb(100,100,100)'
                                }
                            }
                        },
                        states: {
                            hover: {
                                marker: {
                                    enabled: false
                                }
                            }
                        },
                    }
                },
                series: [{
                    data: newData,
                    downsample: {
                        threshold: 0 // 0 disables downsampling
                    },
                    marker: {
                          radius: 2
                    },
                    showInLegend: false,
                    turboThreshold: 0,
                    type: 'scatter',
                }],
                title: {
                    text: plotTitle,
                },
                tooltip: {
                    useHTML: true,
                    formatter: function () {
                        let p = this.point;
                        if (p.protein_id) {
                            return `<b>${p.protein_id}</b><br/>Symbol: ${p.symbol}<br/>Chromosome: ${p.chr}<br/>Location: ${p.pos}<br/>LOD: ${p.LOD}`;

                        } else {
                            return `<b>${p.gene_id}</b><br/>Symbol: ${p.symbol}<br/>Chromosome: ${p.chr}<br/>Location: ${p.pos}<br/>LOD: ${p.LOD}`;
                        }
                    },
                    positioner: function(labelWidth, labelHeight, point) {
                        let tooltipX, tooltipY;
                        let chart = g.chartMediation;

                        if (point.plotX + labelWidth > chart.plotWidth) {
                            tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                        } else {
                            tooltipX = point.plotX + chart.plotLeft + 20;
                        }

                        if (point.plotY + labelHeight > chart.plotHeight) {
                            tooltipY = point.plotY + chart.plotTop - labelHeight - 20;
                        } else {
                            tooltipY = point.plotY + chart.plotTop - 20;
                        }

                        return {
                            x: tooltipX,
                            y: tooltipY
                        };

                    },
                },
                xAxis: {
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return chromosomeAxisText[this.value];
                        }
                    },
                    opposite: true,
                    plotBands: chromosomeBands,
                    tickPositions: chromosomeAxisVals,
                },
                yAxis: {
                    title: {
                        text: 'LOD'
                    }
                },
            });
        }

        /**
         * Start downloading mediation data.
         *
         */
        function generateMediationPlot(id, markerID) {
            let mediateAgainst = $('#dsMediateAgainstID').val();
            let mediateURL = `{{ config.API_R_BASE }}/mediate?dataset=${g.dataSetID}&id=${id}&mid=${markerID}&datasetMediate=${mediateAgainst}`;

            let submitData = {
                urls:[{
                    url_id: 'mediate',
                    url: mediateURL
                }]};

            startTask();

            // reset the chart and clear it
            g.chartMediation = null;
            $('#plotMediation').html('');

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    console.log('data.group_id=', data.group_id);
                    updateMediationData(data.group_id, mediateAgainst);
                },
                error: function() {
                    // TODO: handle
                    alert('Unexpected error');
                }
            });
        }

        /**
         * Update the status of the downloading effect data.
         *
         * @param {string} groupID - the group identifier of the task
         */
        function updateMediationData(groupID, dsMediateAgainst) {
            // send GET request to status URL
            console.log('updateMediationData');
            console.log(groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=True) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function (data, status, request) {
                        console.log('DATA=======', data);
                        if (data['status'] === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data['number_tasks_errors'] !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['error']}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data['number_tasks_errors'] === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['response']['error']}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem performing the Mediation Analysis.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 1 datasets to get

                                    // 1. mediation


                                    let info = data['response_data']['mediate'];
                                    let len = info.response.result.length;
                                    let dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Mediation Points</strong></td><td>${len}</td></tr>
                                            </table>`;
                                    debugMessage('Mediation', dbg);
                                    plotMediation(info['response']['result'], dsMediateAgainst);
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function () {
                                updateMediationData(groupID, dsMediateAgainst);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=True) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                    g.chartMediation = null;
                    $('#plotMediation').html('');
                });
            }
        }



        /**
         * Start downloading effect data.
         *
         */
        function generateEffectPlot(id, chromosome) {
            let effectURL = `{{ config.API_R_BASE }}/foundercoefs?dataset=${g.dataSetID}&id=${id}&chrom=${chromosome}`;
            effectURL += '&blup=' + $('#performBLUP').bootstrapSwitch('state');
            {% if config.API_R_NUM_CORES %}
            effectURL += '&nCores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'effect',
                    url: effectURL
                }]};

            startTask();

            // reset the chart and clear it
            g.chartEffect = null;
            $('#plotEffect').html('');

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    console.log('data.group_id=', data.group_id);
                    updateEffectData(data.group_id);
                },
                error: function() {
                    // TODO: handle
                    alert('Unexpected error');
                }
            });
        }


        /**
         * Update the status of the downloading effect data.
         *
         * @param {string} groupID - the group identifier of the task
         */
        function updateEffectData(groupID) {
            // send GET request to status URL
            console.log('updateEffectData');
            console.log(groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=True) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function (data, status, request) {
                        console.log('DATA=======', data);
                        if (data['status'] === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data['number_tasks_errors'] !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['error']}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data['number_tasks_errors'] === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['response']['error']}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the Effect Plot.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 1 datasets to get

                                    // 1. effect information

                                    let info = data['response_data']['effect'];
                                    let len = info.response.result.length;
                                    let dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Effect Points</strong></td><td>${len}</td></tr>
                                            </table>`;

                                    debugMessage('Effect Plot Information', dbg);

                                    plotEffectData(info['response']['result']);
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function () {
                                updateEffectData(groupID);
                            }, 1000);
                        }
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=True) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                    g.chartEffect = null;
                    $('#plotEffect').html('');
                });
            }
        }

        /**
         * Start downloading LOD Peaks for the dataset.
         *
         * @param {string} dataSetID - the dataSet identifier
         */
        function submitLODPeaks(dataSetID) {
            let lodPeaksURL = '{{ config.API_R_BASE }}/lodpeaks?dataset=' + dataSetID;

            let submitData = {
                urls:[{
                    url_id: 'lodPeaks',
                    url: lodPeaksURL
                }]};

            startTask();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    console.log('data.group_id=', data.group_id);
                    updateProgressPeaks(data.group_id);
                },
                error: function() {
                    // TODO: handle
                    alert('Unexpected error');
                }
            });
        }

        /**
         * Update the status of the downloading LOD Peaks.
         *
         * @param {string} groupID - the group identifier of the task
         */
        function updateProgressPeaks(groupID) {
            // send GET request to status URL
            console.log('updateProgressPeaks');
            console.log(groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=True) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function(data, status, request) {
                        console.log('DATA=======', data);

                        if (data['status'] === 'DONE') {
                            if ('response_data' in data) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    console.log(value);
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (value['response']['error'] + '<br>');
                                    }
                                });
                                console.log(errorMessages);

                                if (errorMessages !== '') {
                                    stopTask();
                                    let message = `Unfortunately, there was a problem getting the LOD Peaks for ${g.dataSets[g.dataSetID].displayName}.`;
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result
                                    console.log('show result');
                                    console.log('using data=', data);
                                    console.log('data.response_data.lodPeaks.response', data['response_data']['lodPeaks']['response']);
                                    setDataSet(g.dataSetID, true, data['response_data']['lodPeaks']['response']['result']);
                                    //progressBar.finish();
                                    stopTask();
                                }
                            }
                            else {
                                // something unexpected happened
                                console.log('very_bad');
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function() {
                                updateProgressPeaks(groupID);
                            }, 1000);
                        }
                    },
                    error: function() {
                        // TODO: handle
                        alert('Unexpected error');
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = '{{ url_for('api.api_cancel', group_id='', _external=True) }}' + groupID;
                // we don't care the response, just try to signal to calce
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                });
            }
        }



        function updateCorrelateDataSet(groupID) {
            // send GET request to status URL
            console.log('updateCorrelateDataSet');
            console.log(groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=True) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function (data, status, request) {
                        console.log('DATA=======', data);
                        if (data['status'] === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data['number_tasks_errors'] !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['error']}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data['number_tasks_errors'] === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['response']['error']}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the Correlation Data.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // 1. Correlation information

                                    console.log('show result');
                                    console.log('using data=', data);
                                    let info = data['response_data']['correlation'];
                                    console.log('info=', info);
                                    let len = info.response.result.length;

                                    let dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Correlation Points</strong></td><td>${len}</td></tr>
                                            </table>`;

                                    debugMessage('Correlation Plot Information', dbg);
                                    displayCorrelation(data.response_data.correlation);

                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function () {
                                updateCorrelateDataSet(groupID);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=True) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                    g.chartCorrelation = null;
                    $('#correlationDataTable').html('');
                    $('#plotCorrelation').html('');
                });
            }

        }

        function switchCorrelationDataSet(dataSetID) {
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            g.chartCorrelation = null;
            g.correlationDatasetID = dataSetID;

            let currentID = '';

            if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                currentID = g.geneID;
            } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                currentID = g.proteinID;
            } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                currentID = g.phenotypeID;
            } else {
                // TODO: handle error
                console.error('MAJOR PROBLEM');
            }

            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${g.dataSetID}&id=${currentID}&datasetCorrelate=${dataSetID}`;

            let submitData = {
                urls:[{
                    url_id: 'correlation',
                    url: urlCorrelation
                }]};

            startTask();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    g.runningTask = true;
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    //let status_url = request.getResponseHeader('Location');
                    //console.log('status_url=', status_url);
                    console.log('data.group_id=', data.group_id);
                    updateCorrelateDataSet(data.group_id);
                },
                error: function() {
                    alert('Unexpected error');
                }
            });




        }

        /**
         * Switch the datasets.
         *
         * @param {string} dataSetID - the dataSet identifier
         */
        function switchDataSet(dataSetID) {
            g.dataSetID = dataSetID;

            let html = `<div class="col">
                            <div class="jumbotron jumbotron-fluid" style="background: none">
                                <div class="container">
                                    <h1 class="display-3">{{ config.MAIN_TITLE|safe }}</h1>
                                    <p class="lead">Please search for a term of interest.</p>
                                </div>
                            </div>
                        </div>`;

            $('#divWelcomeMesage').html(html);
            $('#divItemInfo').addClass('invisible').removeClass('visible');
            $('#divLOD').addClass('invisible').removeClass('visible');
            $('#divSecondRow').addClass('invisible').removeClass('visible');
            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');

            configureCovarFactors();

            // submitLODPeaks calls setDataSet
            submitLODPeaks(dataSetID);
        }

        /**
         * Display the gene information.
         * @param {Object} geneData - gene information
         */
        function setDataSet(dataSetID, force, lodPeaks) {
            console.log('setDataSet=', dataSetID);
            if (!(force) && (QTL.g.dataSetID === dataSetID)) {
                console.log('Same dataset selected, do nothing');
                return;
            }

            if (lodPeaks != null) {
                g.lodPeaks = lodPeaks;
                plotLODPeaks();
            }

            let ds = g.dataSets[dataSetID];
            g.dataSetID = dataSetID;

            console.log('ds=', ds);

            // clear LOD Plot
            // TODO: clear LOD PLOT, Profile PLOT
            g.chartLOD = null;
            $('#lodPlotChart').html('');

            resetSecondaryPlots();

            // make tabs invisible
            $('#navGene').hide();
            $('#navPheno').hide();

            // show the correct tab
            if ((ds.dataType === 'mRNA') || (ds.dataType === 'protein')) {
                $('#navGene').show();
                $('#navGene').tab('show');

                // show some secondary plots
                $('#navPlotMediation').show();

                // reset the screen
                $('#searchResultsDiv').html('');
                $('#searchResultsTableInfo').html('');

                $('#divMediationSelect').html(`Mediating against: ${g.dataSets[dataSetID].displayName}`);
                $('#divMediationSelect').append(`<input type="hidden" id="dsMediateAgainstID" value="${dataSetID}"/>`);

            } else if (ds.dataType === 'phenotype') {
                $('#navPheno').show();
                $('#navPheno').tab('show');

                // hide some secondary plots
                $('#navPlotMediation').hide();


                // hide or show the mediation based upon if there are datasets that can mediate against
                let allDS = [];
                $.each(g.dataSets, function(key, value) {
                    if (value.dataType !== 'phenotype') {
                        allDS.push(value);
                    }
                });

                $('#divMediationSelect').html('');

                if (allDS.length >= 1) {
                    $('#navPlotMediation').show();
                }

                console.log('allDS=', allDS);
                console.log('allDS.length=', allDS.length);

                if (allDS.length === 1) {
                    $('#divMediationSelect').html(`Mediating against: ${allDS[0].displayName}`);
                    $('#divMediationSelect').append(`<input type="hidden" id="dsMediateAgainstID" value="${allDS[0].id}"/>`);
                } else if (allDS.length > 1) {
                    let html = 'Mediate against:';
                    html += '<select id="dsMediateAgainstID">';
                    $.each(allDS, function(idx, elem) {
                        html += `<option value="${elem.id}">${elem.displayName}</option>`;
                    });
                    html += '</select>';

                    $('#divMediationSelect').html(html);
                }


                $('#searchPheno').html('');
                $('#searchPhenoCategoryText').html('');
                $('#searchPhenoCategory').html('');
                $('#searchPhenoResultsInfo').html('');
                $('#searchPhenoResultsDiv').html('');

                let numResults = ds.phenotypes.length;

                // transform the phenotypes into a different structure
                let categories = new Set([]);
                let newPhenotypes = [];
                $.each(ds.phenotypes, function(index, element) {
                    if ((element.isPheno) && (element.isNumeric)) {
                        categories.add(element.category);
                        element.display = element.shortName;
                        newPhenotypes.push({phenotype_id: element.dataName, // use for id
                                            phenotype: element.shortName,   // display on screen
                                            category: element.category,
                                            desc: element.desc});
                    }
                });

                // build the search area
                if (categories.size === 0) {
                    console.log('No categories - ERROR');
                } else {
                    /*
                    $('#searchPheno').html(`<div class="input-group">
                                                <input type="text" id="searchTermPheno" name="searchTermPheno" class="form-control" placeholder="">
                                                <span class="input-group-btn">
                                                    <button id="btnGoPheno" class="btn btn-primary"><i class="fas fa-search"></i></button>
                                                </span>
                                            </div>`);
                    */

                    $('#searchPheno').html('<input type="text" id="searchTermPheno" name="searchTermPheno" class="form-control" placeholder="Please enter a phenotype to filter results...">');

                    // build the phenotypes table
                    $('#searchPhenoResultsDiv').html(`<table id="phenotypesTable" class="table table-striped table-hover table-sm table-bordered">
                                                <thead>
                                                    <th style="background-color: #20a8d8" data-dynatable-sorts="Phenotype">Phenotype</th>
                                                    <th style="display: none">category</th>
                                                    <th style="display: none">desc</th>
                                                </thead>
                                                <tbody>
                                                </tbody>
                                            </table>`);


                    function phenoSort(a, b, attr, direction) {
                        console.log(a,b);
                        // TODO: flush out sort function and maybe also not allow NO sort
                        return 0;//direction > 0 ? comparison : -comparison;
                    }

                    g.phenoDynaTable = $('#phenotypesTable')
                        .bind('dynatable:init', function(e, dynatable) {
                            dynatable.queries.functions['phenotype'] = function(record, queryValue) {
                                if ((record.phenotype.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.desc.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1)) {
                                    return true;
                                }
                            };
                        }).dynatable({
                            dataset: {
                                records: newPhenotypes,
                            },
                            features: {
                                paginate: false,
                                pushState: false,
                                recordCount: true,
                                sort: true,
                                search: false,
                            },
                            inputs: {
                                recordCountPlacement: 'before',
                            },
                            writers: {
                                _rowWriter: function(rowIndex, record, columns, cellWriter) {
                                    return `<tr><td><span class="font-weight-bold"><a href="#" phenotypeid="${record.phenotype_id}">${record.phenotype}</a></span><br><span class="font-italic">${record.category}</span><br><span class="font-weight-light">${record.desc}</span></td></tr>`;
                                }
                            },
                        }).data('dynatable');

                    // build the category select, if more than 1 category
                    categories = Array.from(categories);

                    if (categories.length > 1) {
                        $('#searchPhenoCategoryText').html('Category Filter');

                        let htmlOptions = '<select id="phenoCategorySelect" data-btn-class="btn-primary btn-block" class="form-control">';
                        htmlOptions += '<option>All</option>';
                        $.each(categories, function (i, val) {
                            htmlOptions += `<option>${val}</option>`;
                        });
                        htmlOptions += '</select>';
                        $('#searchPhenoCategory').html(htmlOptions);

                        $('#phenoCategorySelect').extendSelect();
                        $('#phenoCategorySelect').on('change', function () {
                            const selected = $(this).find(':selected').text();
                            console.log('Selected: ' + selected);
                            console.log('phenoDynaTable=', g.phenoDynaTable);
                            if (selected === "All") {
                                g.phenoDynaTable.queries.remove("category");
                            } else {
                                g.phenoDynaTable.queries.add("category", selected);
                            }

                            g.phenoDynaTable.process();
                            $('#phenotypesTable tbody a').on('click', function (evt) {
                                evt.preventDefault();
                                let that = $(this);
                                selectPhenotype(that.attr('phenotypeid'), g.dataSetID);
                                return false;
                            });
                        });
                    }

                    $('#searchTermPheno').keyup(function (event) {
                        if ( event.which === 13 ) {
                            event.preventDefault();
                        }
                        let q = $(this).val();
                        //console.log('q=', q);
                        if (q === "") { q = undefined; }
                        if (q) {
                            console.log('adding', q);

                            g.phenoDynaTable.queries.add('phenotype', q);
                            console.log(g.phenoDynaTable.queries);
                        } else {
                            console.log('removing');
                            g.phenoDynaTable.queries.remove('phenotype');
                            console.log(g.phenoDynaTable.queries);

                        }

                        console.log('g.phenoDynaTable.process()');
                        g.phenoDynaTable.process();
                        $('#phenotypesTable tbody a').on('click', function (evt) {
                            evt.preventDefault();
                            let that = $(this);
                            selectPhenotype(that.attr('phenotypeid'), g.dataSetID);
                            return false;
                        });

                    });

                    $('#phenotypesTable tbody a').on('click', function (evt) {
                        evt.preventDefault();
                        let that = $(this);
                        selectPhenotype(that.attr('phenotypeid'), g.dataSetID);
                        return false;
                    });
                }

            } else {
                // TODO: handle error
                console.log('ERROR in setDataSet:', ds);
            }
        }

        function exportProfileData(id, data, factors) {
            let csvContent = '"mouse_id","expression"';
            $.each(factors, function(k, v) {
                csvContent += `,"${v['display.name']}"`;
            });
            csvContent += '\n';

            $.each(data.data, function(k, v) {
                let line = `"${v.mouse_id}",${v.expression}`;
                $.each(factors, function(fk, fv) {
                    line += `,"${v[fv['column.name']]}"`;
                });

                csvContent += `${line}\n`;
            });

            downloadCSV(csvContent, `${id}_PROFILE.csv`, 'text/csv;encoding:utf-8');
        }


        /**
         * Plot profile data.
         * @param {string} id - the unique identifier
         */
        function plotProfile(id) {
            console.log('plotProfile(', id, ')');
            // profiles are covar_factors
            let data = g.expressionData;
            let categories = getFactorOrder();
            console.log('categories=', categories);

            let categorySeries = $('#factorSeries').find(':selected').val();
            console.log('categorySeries=', categorySeries);

            // using this so plots have same jitter
            Math.seedrandom('profileplot');

            let newData = {};
            let allCategories = [];
            let cfMap = {};

            $.each(g.dataSets[g.dataSetID].covarFactors, function(idx, elem) {
                cfMap[elem['column.name']] = elem['display.name'];
            });

            console.log('cfMap=', cfMap);


            let title = [];

            $.each(categories, function(idx, elem) {
                newData[elem] = [];
                allCategories.push(data.dataTypes[elem]);
                console.log(idx, elem);
                title.push(cfMap[elem]);
            });


            let chartTitle = title.join(' x ');
            allCategories = permutateArrays(allCategories);

            console.log('All Categories=', allCategories);

            // this will give us 9 points per "category"
            let jitter = 5;
            let jitterPad = 2;
            let point = [];
            let pointToLabel = {};
            let start = jitterPad + jitter + 1;
            let stepAmount = (jitterPad * 2) + (jitter * 2) + 1;

            let newCategories = {};
            let newCategoriesOrder = [];

            console.log("Loop through categories...");
            $.each(allCategories, function(idx, elem) {
                let catName = elem.join(':');
                let combo = {};
                $.each(elem, function(i, e) {
                    combo[categories[i]] = e;
                });
                newCategories[catName] = {
                    id: catName,
                    categoryValues: elem,
                    combo: combo,
                    x: idx,
                    start: start + (idx * stepAmount),
                    values: []
                };

                newCategoriesOrder.push(catName);
                point.push(start + (idx * stepAmount));
                pointToLabel[start + (idx * stepAmount)] = catName;
            });

            console.log('point=', point);
            console.log('pointToLabel=', pointToLabel);

            let seriesData = {};
            $.each(data.dataTypes[categorySeries], function(idx, val) {
                seriesData[val] = { values: [] };
            });

            console.log('seriesData=', seriesData);

            console.log('Looping through data.data...');
            $.each(data.data, function(idx, val) {
                //console.log('Looping through ', data.dataTypes[categorySeries]);
                $.each(data.dataTypes[categorySeries], function(i, v) {
                    //console.log(idx, val, i, v, categorySeries);

                    let abc = [];
                    $.each(categories, function(_i, _elem) {
                        abc.push(val[_elem]);
                    });
                    let catName = abc.join(':');

                    if (val['expression'] === undefined) {
                        //console.log('Undefined data=', val);
                    } else {

                        newCategories[catName].values.push(val['expression']);

                        // put the data in the correct buckets

                        if (v === val[categorySeries]) {
                            seriesData[v].values.push({
                                type: 'scatter',
                                val: val,
                                x: newCategories[catName].start + getRandomInt(-1 * jitter, jitter),
                                y: val['expression']
                            });
                        }
                    }
                });
            });

            // convert from object, to array

            let series = [];
            let colors = ['#7cb5ec', '#ed3215', '#8085e9', '#f7a35c', '#90ed7d',
                '#e4d354', '#2b908f', '#f45b5b', '#91e8e1'];
            $.each(seriesData, function(key, val) {
                /*
                In the old code Mark Keller wanted to differentiate by color,
                male and female mice.  I've removed but left code in.

                let color = null;

                if (categorySeries.toLowerCase() === 'sex') {
                    if (key.toLowerCase().charAt(0) === 'm') {
                        color = '#00BFC4';
                    } else if (key.toLowerCase().charAt(0) === 'f') {
                        color = '#F8766D';
                    }
                }
                */

                series.push({
                    data: val.values,
                    name: key,
                    marker: {
                        radius: 2
                    },
                    color: colors[series.length % colors.length]
                });
            });

            let boxSeries = [];

            $.each(newCategories, function(key, val) {
                if (val.values.length > 0) {
                    let boxValues = getBoxValues(val.values);
                    boxValues.x = val.start;
                    boxValues.name = val.id;
                    boxValues.enableMouseTracking = false;
                    boxSeries.push(boxValues);
                }
            });

            series.push({
                type: 'boxplot',
                data: boxSeries,
                showInLegend: false,
            });

            let exporting = appendExportButton('Download CSV', function() {
                exportProfileData(id, data, g.dataSets[g.dataSetID].covarFactors);
            });
            exporting.filename = id + '_PROFILE';

            Highcharts.chart({
                chart: {
                    type: 'scatter',
                    renderTo: 'profilePlotChart',
                },

                title: {
                    text: chartTitle,
                    verticalAlign: 'bottom'
                },

                xAxis: {
                    // min and max forces the "categories" to be shown even when no data is available
                    min: 0,
                    max: point[point.length-1] + jitter + jitterPad,
                    tickPositions: point,
                    labels: {
                        formatter: function () {
                            return pointToLabel[this.value];
                        },
                        rotation: -45
                    }

                },

                yAxis: {
                    title: {
                        text: id
                    }
                },

                series: series,

                exporting: exporting,

                plotOptions: {
                    boxplot: {
                        color: '#000000',
                        enableMouseTracking: false,
                        lineWidth: 1,
                        medianColor: '#000000',
                        medianWidth: 1,
                        stemColor: '#000000',
                        stemWidth: 1,
                        whiskerColor: '#000000',
                        whiskerLength: '50%',
                        whiskerWidth: 1
                    },
                    scatter: {
                        tooltip: {
                            headerFormat: null,
                            pointFormatter: function() {
                                let val = this.options.val;
                                return `<b>${val.mouse_id}</><br><p>${val.expression}</p>`;
                            }
                        }
                    }
                }
            });

        }

        /**
         * Perform gene search.
         */
        function findGene(searchVal) {
            let button = $('#btnGo');
            let term = $('#searchTerm');
            term.disable(true);
            //button.button('loading');

            if (searchVal.length === 0) {
                term.focus();
                term.disable(false);
                //button.button('reset');
                return;
            }

            let options = {
                species: g.speciesID,
                version: g.ensemblVersion,
                limit: 100
            };

            console.log(options);

            startTask();
            g.ensimpl.search(searchVal, options, searchCallback);
        }

        /**
         * Populate the search results.
         */
        function searchCallback() {
            console.log(g.ensimpl);

            if (g.ensimpl.response.result.matches === null) {
                $('#searchResultsDiv').html('');
                $('#btnGo').button('reset');
                $('#searchTerm').disable(false);
                $('#searchResultsTableInfo').html('No results found');
                stopTask();
                return;
            }

            let tbl = '<table id="searchResultsTable" class="table table-striped table-hover table-sm table-bordered">';
            tbl += '<thead>';
            tbl += '<tr>';
            tbl += '<th class="" scope="col">ID</th>';
            tbl += '<th class="" scope="col">Symbol</th>';
            /*
            tbl += '<th class="d-none d-md-table-cell" scope="col">Position</th>';
            tbl += '<th class="d-none d-xl-table-cell" scope="col">Match Reason</th>';
            tbl += '<th class="d-none d-xl-table-cell" scope="col">Description</th>';
            */
            tbl += '</tr>';
            tbl += '</thead>';
            tbl += '<tbody id="tblBody">';
            tbl += '</tbody>';
            tbl += '</table>';
            $('#searchResultsDiv').html(tbl);

            let currentDataSet = g.dataSets[g.dataSetID];
            let response = g.ensimpl.response;
            let searchResults = {};
            /*
            g.searchTable = $('#searchResultsTable').DataTable({
                //data: response.result.matches,
                paging: false,
                order: [],   // prevent automatic ordering
                scrollY: 200,
                searching: false,
                lengthChange: false,
                scrollCollapse: true,
                responsive: true,
                //columns: columns
            });
            */

            $.each(response.result.matches, function(idx, match) {


                let row = '<tr>';
                searchResults[match.ensembl_gene_id] = match;

                if (currentDataSet.dataType === 'mRNA') {
                    if (currentDataSet.gene_ids[match.ensembl_gene_id]) {
                        row += '<td class="text-nowrap"><i id="matchInfo" geneID="' + match.ensembl_gene_id + '" class="fas fa-info-circle"></i> <a href="#" geneID="' + match.ensembl_gene_id + '">' + match.ensembl_gene_id + '</a></td>';
                    } else {
                        row += '<td class="text-nowrap"><i id="matchInfo" geneID="' + match.ensembl_gene_id + '" class="fas fa-info-circle"></i> ' + match.ensembl_gene_id + '</td>';
                    }
                } else if (currentDataSet.dataType === 'protein') {
                    if (currentDataSet.gene_ids[match.ensembl_gene_id]) {
                        let ps = [];
                        row += '<td><div class="text-nowrap"><i id="matchInfo" geneID="' + match.ensembl_gene_id + '" class="fas fa-info-circle"></i> ' + match.ensembl_gene_id + '</data>';
                        row += '<div style="padding-left: 20px;">';

                        for (i in currentDataSet.gene_ids[match.ensembl_gene_id].protein_ids) {
                            let p = currentDataSet.gene_ids[match.ensembl_gene_id].protein_ids[i];
                            ps.push('<em><a href="#" geneID="' + match.ensembl_gene_id + '" proteinID="' + p + '">' + p + '</a></em>');
                        }

                        row += ps.join('<br>');
                        row += '</div></td>';
                    } else {
                        row += '<td class="text-nowrap"><i id="matchInfo" geneID="' + match.ensembl_gene_id + '" class="fas fa-info-circle"></i> ' + match.ensembl_gene_id + '</td>';
                    }
                } else {
                    // TODO: handle error
                    console.error('THIS SHOULD NOT HAPPEN');
                }

                row += '<td>' + match.symbol + '</td>';
                /*
                row += '<td class="d-none d-md-table-cell">' + match.symbol + '</td>';
                row += '<td class="d-none d-xl-table-cell">' + match.symbol + '</td>';
                row += '<td class="d-none d-xl-table-cell">' + match.symbol + '</td>';
                */
                row += '</tr>';

                $('#tblBody').append(row);
            });



            if (response.result.num_results === 1) {
                $('#searchResultsTableInfo').html(response.result.matches.length.toLocaleString() + ' result');
            } else if (response.result.matches.num_matches < 100) {
                $('#searchResultsTableInfo').html(response.result.matches.length.toLocaleString() + ' results');
            } else {
                $('#searchResultsTableInfo').html('Showing first ' + response.result.matches.length.toLocaleString() + ' of ' + response.result.num_results.toLocaleString() + ' results');
            }

            $('#searchResultsTable a').on('click', function (evt) {
                evt.preventDefault();
                let that = $(this);
                //let regressLocal = $("#cond_local").is(":checked");

                selectGeneProtein(that.attr('geneid'), that.attr('proteinid'), g.dataSetID);
                return false;
            });

            setTimeout(function () {
                $('[id="matchInfo"]').each(function (idx, elem) {
                    let that = $(this);
                    $(this).popover({
                        placement: 'right',
                        html: true,
                        trigger: 'hover',
                        container: 'body',
                        //template: '<div class="popover container my-custom-class" role="tooltip"><div class="arrow"></div><h3 class="popover-title"></h3><div class="popover-content"></div></div>',
                        content: function () {
                            let d = searchResults[elem.getAttribute("geneid")];
                            console.log(d);
                            let h = '<table class="table table-sm">';
                            h += '<tr><td><strong>Symbol</strong> </td><td>' + d['symbol'] + '</td></tr>';
                            h += '<tr><td><strong>Location</strong> </td><td>' + d['chromosome'] + ':' + d['position_start'] + '-' + d['position_end'] + '</td></tr>';
                            h += '<tr><td><strong>Name</strong> </td><td>' + d['name'] + '</td></tr>';
                            h += '<tr><td><strong>Synonyms</strong> </td><td>';

                            let s = d['synonyms'];
                            if (s) {
                                if (s.length > 5) {
                                    let s2 = s.slice(1, 6);
                                    h += s2.join('<br>');
                                    h += '<br><i>(' + (s.length - s2.length) + ' more synonyms)</i>';
                                } else {
                                    h += s.join('<br>');
                                }
                            }
                            h += '</td></tr>';
                            h += '<tr><td><strong>Match Reason</strong> </td><td>' + d['match_reason'] + '</td></tr>';
                            h += '<tr><td><strong>Match Value</strong> </td><td>' + d['match_value'] + '</td></tr>';
                            h += '</table>';

                            return h;
                        }
                    });
                });

            });

            let dbg = `<table class="table table-sm">
                    <tr><td><strong>Number Matches</strong></td><td>${g.ensimpl.response.result.num_matches}</td></tr>
                    <tr><td><strong>Number Results</strong></td><td>${g.ensimpl.response.result.num_results}</td></tr>
                    </table>`;
            debugMessage('Search Information', dbg);


            //debugMessage();

            g.searchResults = searchResults;
            //g.searchResultsLevel = currentLevel;
            $('#btnGo').button('reset');
            $('#searchTerm').disable(false);

            stopTask();

            // simulate a click event if the search yields only 1 result
            if ($('#searchResultsTable a').length === 1) {
                $('#searchResultsTable a')[0].click();
            }
        }

        function generateSecondaryPlot(plot, id, markerID, chromosome, location) {
            console.log('Generating Secondary Plot')
            if (plot === 'navPlotMediation') {
                generateMediationPlot(id, markerID);
            } else if (plot === 'navPlotEffect') {
                //generateEffectPlot(id, chromosome, $("#mediate_blup").is(':checked'), $("#cond_local").is(":checked"));
                generateEffectPlot(id, chromosome);
            } else if (plot === 'navPlotSNPs') {
                generateSNPAssocPlot(id, chromosome, location);
            } else {
                console.error('Unknown plot ', plot);
            }
        }

        // The download function takes a CSV string, the filename and mimeType as parameters
        // Scroll/look down at the bottom of this snippet to see how download is called
        function downloadCSV(content, fileName, mimeType) {
            let a = document.createElement('a');
            mimeType = mimeType || 'application/octet-stream';

            if (navigator.msSaveBlob) { // IE10
                navigator.msSaveBlob(new Blob([content], {
                    type: mimeType
                }), fileName);
            } else if (URL && 'download' in a) { //html5 A[download]
                a.href = URL.createObjectURL(new Blob([content], {
                    type: mimeType
                }));
                a.setAttribute('download', fileName);
                document.body.appendChild(a);
                a.click();
                document.body.removeChild(a);
            } else {
                location.href = 'data:application/octet-stream,' + encodeURIComponent(content); // only this mime type is supported
            }
        }


        function exportLODChartData(id, data) {
            let csvContent = '"id","chromosome","position","lod"\n';
            $.each(data, function(k, v) {
                csvContent += `"${v.id}","${v.chr}",${v.chrPos},${v.y}\n`;
            });

            downloadCSV(csvContent, `${id}_LOD.csv`, 'text/csv;encoding:utf-8');

        }

        function plotLODChart(dataLOD) {
                // "1_3000000", "1", 3, 1.4975
            let dataByChrom = {};
            let maxLOD = -Infinity;
            let minLOD = Infinity;
            let xAxisTitle = '';
            let newLodData = [];
            let currentID = null;

            console.log('dataLOD=', dataLOD);

            if (dataLOD !== null) {
                maxLOD = dataLOD[0][3];
                minLOD = -2;

                if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                    currentID = g.geneID;
                    xAxisTitle = g.geneID + ' (' + g.currentGene.gene[g.geneID].symbol + ')';
                } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                    currentID = g.proteinID;
                    xAxisTitle = g.proteinID + ' (' + g.currentGene.gene[g.geneID].symbol + ')';
                } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                    // pheno
                    xAxisTitle = g.phenotypeID;
                    currentID = g.phenotypeID;
                    if (currentID !== g.dataSets[g.dataSetID].phenotypes[currentID].shortName) {
                        xAxisTitle += (' (' + g.dataSets[g.dataSetID].phenotypes[currentID].shortName + ')');
                    }
                } else {
                    // TODO: handle error
                    console.error('MAJOR PROBLEM');
                }

                $.each(dataLOD, function(index, element) {
                    //
                    // element = {"[id]","[chr]",[position],[data]}
                    //
                    if (element[1] in g.chromosomes.chr) {

                        let chr = g.chromosomes.chr[element[1]];
                        let x = Math.round(chr.start + (element[2] * 1000000.0));
                        let y = element[3];

                        let d = {
                            x: x,
                            y: y,
                            id: element[0],
                            text: element[3],
                            chr: chr.chromosome,
                            chrPos: element[2],
                        };

                        newLodData.push(d);

                        if (element[3] > maxLOD) {
                            maxLOD = element[3];
                        }
                    }
                });

                maxLOD += 5;
            }

            newLodData.sort(compareX);

            console.log('newLodData=', newLodData);

            //
            // build the vertical chromosome bars
            //
            let chromosomeAxisVals = [];
            let chromosomeAxisText = {};
            let chromosomeBands = [];

            $.each(g.chromosomes.idx, function(index, element) {

                chromosomeAxisVals.push(element.mid);
                chromosomeAxisText[element.mid] = element.chromosome;

                // TODO: style this
                let fillColor = '#eeeeee';
                if ((index % 2) === 1) {
                    chromosomeBands.push({
                        color: fillColor,
                        from: element.start,
                        to: element.end
                    });

                    console.log(fillColor);
                }
            });

            let plotShapes = [];
            let plotAnnotations = [];

            {% if config.PLOT_LOD_LINES %}
            //
            // build the horizontal lines
            //
            let lineVals = [
                {% for l in config.PLOT_LOD_LINES %}
                    {'val':{{l.val}}, 'color':'{{l.color}}', 'text':'{{l.text }}' },
                {% endfor %}
            ];
            let _plotLines = [
                {% for l in config.PLOT_LOD_LINES %}
                    {value:{{l.val}}, color:'{{l.color}}', zIndex:5, width:2, label:{
                        text:'{{l.text }}', align:'right'
                    } },
                {% endfor %}
                ];
            for (let l in lineVals) {
                plotShapes.push({
                    type: 'line',
                    xref: 'paper',
                    x0: 0,
                    y0: lineVals[l].val,
                    x1: 1,
                    y1: lineVals[l].val,
                    line: {
                        color: lineVals[l].color,
                        width: 1.5
                    },
                    layer: 'above',
                });

                plotAnnotations.push({
                    x: 1,
                    y: lineVals[l].val,
                    xref: 'paper',
                    yref: 'y',
                    text: lineVals[l].text,
                    xanchor: 'left',
                    yanchor: 'middle',
                    showarrow: false,
                });
            }
            {% endif %}


            let exporting = appendExportButton('Download CSV', function() {
                exportLODChartData(currentID, newLodData);
            });
            exporting.filename = currentID + '_LOD';

            g.chartLOD = Highcharts.chart({
                boost: {
                    debug: {
                        showSkipSummary: true
                    },
                    enabled: true,
                    useGPUTranslations: true,
                },
                chart: {
                    alignTicks: false,
                    renderTo: 'lodPlotChart',
                    zoomType: 'x',
                    events: {
                        click: function(event) {
                            if (event.target.tagName === 'tspan') {
                                // do not generate secondary plot for user clicking "Reset zoom"
                                return;
                            }
                            let plot = $('#secondaryPlotTabs').find('.active').attr('id');
                            let markerID = this.hoverPoint.id;
                            let chromosome = this.hoverPoint.chr;
                            let location = this.hoverPoint.chrPos;

                            g.secondaryPlotMarkerID = markerID;
                            g.secondaryPlotChromosome = chromosome;
                            g.secondaryPlotLocation = location;

                            generateSecondaryPlot(plot, currentID, markerID, chromosome, location)
                        }
                    }
                },
                exporting: exporting,
                plotOptions: {
                    series: {
                        findNearestPointBy: 'xy',
                        cursor: 'pointer',
                        events: {
                            click: function(event) {
                                let plot = $('#secondaryPlotTabs').find('.active').attr('id');
                                let markerID = event.point.id;
                                let chromosome = event.point.chr;
                                let location = event.point.chrPos;

                                g.secondaryPlotMarkerID = markerID;
                                g.secondaryPlotChromosome = chromosome;
                                g.secondaryPlotLocation = location;

                                generateSecondaryPlot(plot, currentID, markerID, chromosome, location);
                            }
                        }
                    }
                },

                series: [{
                    boostThreshold: 1,
                    color: 'black',
                    data: newLodData,
                    lineWidth: 0.5,
                    marker: {
                       enabled: false
                    },
                    showInLegend: false,
                    turboThreshold: 0,
                    type: 'line',
                }],
                title: {
                    text: xAxisTitle
                },
                tooltip: {
                    positioner: function(labelWidth, labelHeight, point) {
                        let tooltipX, tooltipY;
                        let chart = g.chartLOD;

                        if (point.plotX + labelWidth > chart.plotWidth) {
                            tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                        } else {
                            tooltipX = point.plotX + chart.plotLeft + 20;
                        }

                        if (point.plotY + labelHeight > chart.plotHeight) {
                            tooltipY = point.plotY + chart.plotTop - labelHeight - 20;
                        } else {
                            tooltipY = point.plotY + chart.plotTop - 20;
                        }

                        return {
                            x: tooltipX,
                            y: tooltipY
                        };
                    },
                    useHTML: true,
                    formatter: function () {
                        let p = this.point;
                        return `<b>${p.id}</b><br/>LOD: ${p.y}<br/>Chromosome: ${p.chr}<br/>Position: ${p.chrPos}`;
                    }
                },
                xAxis: {
                    opposite: true,
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return chromosomeAxisText[this.value];
                        }
                    },
                    tickPositions: chromosomeAxisVals,
                    plotBands: chromosomeBands,

                },
                yAxis: {
                    title: {
                        text: 'LOD'
                    },
                    {% if config.PLOT_LOD_LINES %}
                    plotLines: _plotLines
                    {% endif %}
                },
            });
        }

        /**
         * Get the selected factors.
         * @returns {Array} an array of strings
         */
        function getFactorOrder() {
            let selected = [];
            $('#factor-view-select option:selected').each(function() {
                selected.push([$(this).val(), $(this).attr('order')]);
            });

            selected.sort(function(a, b) {
                return a[1] - b[1];
            });

            let ordered = [];
            for (let i = 0; i < selected.length; i++) {
                ordered.push(selected[i][0]);
            }

            return ordered;
        }

        function configureCovarFactors() {
            // current dataset
            let ds = g.dataSets[g.dataSetID];

            let orderCount = ds.covarFactors.length;
            g.orderCount = ds.covarFactors.length;

            $('#factor-view-select').multiselect({
                onChange: function(option, checked) {
                    if (checked) {
                        g.orderCount++;
                        console.log('g.orderCount=', g.orderCount);
                        $(option).attr('order', g.orderCount);
                    }
                    else {
                        $(option).attr('order', '');
                    }
                },
                maxHeight: 400,
                dropUp: false,
                buttonWidth: '100%',
                buttonClass: 'btn btn-secondary',
                buttonText: function(options) {
                    if (options.length === 0) {
                        return 'None selected';
                    }
                    else {
                        let selected = [];

                        options.each(function() {
                            selected.push([$(this).text(), $(this).attr('order')]);
                        });

                        selected.sort(function(a, b) {
                            return a[1] - b[1];
                        });

                        let text = '';
                        for (let i = 0; i < selected.length; i++) {
                            text += ((i+1) + ': ' + selected[i][0] + ', ');
                        }

                        return text.substr(0, text.length - 2);
                    }
                }
            });

            let selectOptions = [];
            let html = '<select id="factorSeries">';
            let htmlCorrelation = '<select id="factorSeriesCorrelation">';

            console.log('ds.covarFactors=', ds.covarFactors);

            $.each(ds.covarFactors, function(index, element) {
                console.log('ds.covarFactor, element=', element);
                selectOptions.push({
                    label: element['display.name'],
                    order: index,
                    selected: true,
                    title: element['display.name'],
                    value: element['column.name'],
                });
                html += `<option value="${element['column.name']}">${element['display.name']}</option>`
                htmlCorrelation += `<option value="${element['column.name']}">${element['display.name']}</option>`
            });
            html += '</select>';
            htmlCorrelation += '</select>';

            $('#factor-view-select').multiselect('dataprovider', selectOptions);

            $('#divFactorSeries').html(html);
            $('#divFactorSeriesCorrelation').html(htmlCorrelation);

            $('#factorSeries').extendSelect();
            $('#factorSeries').on('change', function () {
                if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                    plotProfile(g.geneID);
                } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                    plotProfile(g.proteinID);
                } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                    plotProfile(g.phenotypeID);
                } else {
                    // TODO: major error
                    console.error('MAJOR ERROR');
                }
            });

            $('#factorSeriesCorrelation').extendSelect();
            $('#factorSeriesCorrelation').on('change', function (evt) {
                console.log(evt.target.value);
                plotCorrelationChart();
            });
        }

        function formatStrToNum(val, decimals) {
            return Number.parseFloat(val).toFixed(decimals);
        }


        function displayCorrelation(correlationData) {
            $('#correlationDataTable').html('');
            console.log('correlationData=', correlationData);
            g.correlationData = correlationData.response.result;

            let htmlBody = '<table id="corTable" class="table table-sm table-sm-text">';

            if (g.dataSets[g.correlationDatasetID].dataType === 'mRNA') {
                htmlBody += '<tr><td><strong>Value</strong></td><td><strong>Symbol</strong></td><td><strong>Location</strong></td></tr>';
                $.each(correlationData.response.result, function(idx, value) {
                    if (idx < 1000) {
                        htmlBody += '<tr><td>' + formatStrToNum(value.cor, 4) + `</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${value.symbol}</a></td><td>${value.chr}:` + formatStrToNum(value.start, 2) + '-' + formatStrToNum(value.end, 2) + '</td></tr>';
                    }
                });
            } else if (g.dataSets[g.correlationDatasetID].dataType === 'protein') {
                htmlBody += '<tr><td><strong>Value</strong></td><td><strong>Symbol</strong></td><td><strong>Chr</strong></td></tr>';
                $.each(correlationData.response.result, function(idx, value) {
                    if (idx < 1000) {
                        htmlBody += '<tr><td>' + formatStrToNum(value.cor, 4) + `</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${value.symbol}</a></td><td>${value.chr}:` + formatStrToNum(value.start, 2) + '-' + formatStrToNum(value.end, 2) + '</td></tr>';
                    }
                });
            } else if (g.dataSets[g.correlationDatasetID].dataType === 'phenotype') {
                htmlBody += '<tr><td><strong>Value</strong></td><td><strong>ID</strong></td></tr>';
                $.each(correlationData.response.result, function(idx, value) {
                    if (idx < 1000) {
                        htmlBody += '<tr><td>' + formatStrToNum(value.cor, 4) + `</td><td><a href="#" dSymbol="" dID="${value.id}">${value.id}</a></td></tr>`;
                    }
                });
            } else {
                // TODO: major error
                console.error('MAJOR ERROR');
            }

            htmlBody += '</table>';
            $('#correlationDataTable').html(htmlBody);

            $('#corTable a').click(function(evt) {
                evt.preventDefault();
                let that = $(this);
                let cid = null;

                if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                    cid = g.geneID;
                } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                    cid = g.proteinID;
                } else if (g.dataSets[g.dataSetID].dataType === 'phenotype') {
                    cid = g.phenotypeID;
                }

                generateCorrelationPlot(g.dataSetID, cid, $("#correlationDatasetSelect").val(), that.attr('dID'), that.attr('dSymbol'));
                return false;
            });
        }

        function exportCorrelationPlotData(id, idCorrelate, data) {
            let csvContent = `"mouse_id","${id}","${idCorrelate}"\n`;

            $.each(data, function(k, v) {
                let line = `"${v['mouse.id']}",${v['x']},${v['y']}`;
                csvContent += `${line}\n`;
            });

            downloadCSV(csvContent, `${id}_${idCorrelate}_MEDIATION.csv`, 'text/csv;encoding:utf-8');
        }


        function plotCorrelationChart() {
            let data = g.correlationChartData;
            console.log('plotCorrelationChart, data=', data);
            let plotTitle = '';

            let categorySeries = $('#divFactorSeriesCorrelation').find(':selected').val();
            console.log('categorySeries=', categorySeries);

            let seriesData = {};

            // need to pass  back factors or use global?
            $.each(data.dataTypes[categorySeries], function(idx, val) {
                seriesData[val] = { values: [] };
            });

            let minValue = Infinity;
            let maxValue = -Infinity;

            $.each(data.data, function(idx, element) {
                if ((element['x'] !== undefined) && (element['y'] !== undefined)) {
                    minValue = Math.min(minValue, element['x']);
                    minValue = Math.min(minValue, element['y']);
                    maxValue = Math.max(maxValue, element['x']);
                    maxValue = Math.max(maxValue, element['y']);

                    //console.log('Looping through ', data.dataTypes[categorySeries]);
                    $.each(data.dataTypes[categorySeries], function (i, v) {
                        // put the data in the correct buckets
                        if (v === element[categorySeries]) {
                            seriesData[v].values.push(element);
                        }
                    });
                }
            });

            //console.log('seriesData=', seriesData);

            // convert from object, to array

            let series = [];
            let colors = ['#7cb5ec', '#ed3215', '#8085e9', '#f7a35c', '#90ed7d',
                '#e4d354', '#2b908f', '#f45b5b', '#91e8e1'];
            $.each(seriesData, function(key, val) {
                series.push({
                    data: val.values,
                    name: key,
                    marker: {
                        radius: 2
                    },
                    type: 'scatter',
                    color: colors[series.length % colors.length]
                });
            });

            minValue = Math.floor(minValue);
            maxValue = Math.ceil(maxValue);

            console.log('minValue=', minValue);
            console.log('maxValue=', maxValue);

            console.log('series=', series);

            let yAxisTitle = data.idCorrelate;

            /*
            if (correlateSymbol !== '') {
                yAxisTitle = correlateSymbol;
            }
            */

            let exporting = appendExportButton('Download CSV', function() {
                exportCorrelationPlotData(data.id, data.idCorrelate, data.data);
            });
            exporting.filename = data.id + '_' + data.idCorrelate + '_CORRELATION';

            g.chartCorrelation = Highcharts.chart({
                chart: {
                    renderTo: 'plotCorrelation',
                    zoomType: 'xy',
                    height: 350,
                    width: 350
                },
                exporting: exporting,
                plotOptions: {
                    scatter: {
                        marker: {
                            radius: 5,
                            states: {
                                hover: {
                                    enabled: true,
                                    lineColor: 'rgb(100,100,100)'
                                }
                            }
                        },
                        states: {
                            hover: {
                                marker: {
                                    enabled: false
                                }
                            }
                        },
                    }
                },
                /*
                series: [{
                    data: newData,
                    downsample: {
                        threshold: 0 // 0 disables downsampling
                    },
                    marker: {
                          radius: 2
                    },
                    showInLegend: false,
                    turboThreshold: 0,
                    type: 'scatter',
                }],
                */
                series: series,
                title: {
                    text: '',
                },
                tooltip: {
                    useHTML: true,
                    formatter: function () {
                        let p = this.point;
                        return `<b>${p['mouse.id']}</b>`;
                    },
                    positioner: function(labelWidth, labelHeight, point){
                        let tooltipX, tooltipY;
                        let chart = g.chartCorrelation;
                        if (point.plotX + labelWidth > chart.plotWidth) {
                            tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                        } else {
                            tooltipX = point.plotX + chart.plotLeft + 20;
                        }
                        tooltipY = point.plotY + chart.plotTop - 20;
                        return {
                            x: tooltipX,
                            y: tooltipY
                        };
                    },
                },
                xAxis: {
                    endOnTick: true,
                    gridLineWidth: 1,
                    min: minValue,
                    max: maxValue,
                    //tickAmount: 4,
                    title: {
                        text: data.id
                    },
                },
                yAxis: {
                    endOnTick: true,
                    gridLineWidth: 1,
                    min: minValue,
                    max: maxValue,
                    //tickAmount: 4,
                    title: {
                        text: yAxisTitle
                    }
                },
            });

        }

        function updateCorrelationData(groupID, currentDataset, currentID, correlateDataset, correlateID, correlateSymbol) {
            // send GET request to status URL
            console.log('updateCorrelationData');
            console.log(groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=True) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function (data, status, request) {
                        console.log('DATA=======', data);
                        if (data['status'] === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data['number_tasks_errors'] !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['error']}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data['number_tasks_errors'] === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['response']['error']}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the Correlation Plot.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // 1. Correlation information

                                    console.log('show result');
                                    console.log('using data=', data);
                                    let info = data['response_data']['correlationPlot'];
                                    let len = info.response.result.data.length;

                                    let dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Correlation Points</strong></td><td>${len}</td></tr>
                                            </table>`;

                                    debugMessage('Correlation Plot Information', dbg);

                                    g.correlationChartData = data.response_data.correlationPlot.response.result;

                                    plotCorrelationChart();
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function () {
                                updateCorrelationData(groupID, currentDataset, currentID, correlateDataset, correlateID, correlateSymbol);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=True) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                    g.chartCorrelation = null;
                    $('#correlationDataTable').html('');
                    $('#plotCorrelation').html('');
                });
            }

        }

        function generateCorrelationPlot(currentDataset, currentID, correlateDataset, correlateID, correlateSymbol) {
            let urlCorrelationPlot = `{{ config.API_R_BASE }}/correlationPlotData?dataset=${currentDataset}&id=${currentID}&datasetCorrelate=${correlateDataset}&idCorrelate=${correlateID}`;

            let submitData = {
                urls:[{
                    url_id: 'correlationPlot',
                    url: urlCorrelationPlot
                }]};


            startTask();

            // reset the chart and clear it
            g.correlationPlot = null;
            $('#plotCorrelation').html('');


            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    console.log('data.group_id=', data.group_id);
                    updateCorrelationData(data.group_id, currentDataset, currentID, correlateDataset, correlateID, correlateSymbol);
                },
                error: function() {
                    // TODO: handle
                    alert('Unexpected error');
                }
            });

        }


        function resetSecondaryPlots() {
            g.chartMediation = null;
            g.chartEffect = null;
            g.chartSNPAssoc = null;
            $('#plotMediation').html('');
            $('#plotEffect').html('');
            $('#plotSNPAssoc').html('');
            $('#snpWindowMenu').html('');
        }

        function updateGeneSelect(groupID) {
            // send GET request to status URL
            console.log('update progress, calling .ajax');

            console.log(groupID);

            if (g.runningTask) {
                let statusURL = `{{ url_for('api.api_status', group_id='', _external=True) }}${groupID}`;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function (data, status, request) {
                        console.log('DATA=======', data);

                        /* A response will be like this:

                           {
                            number_tasks_completed: 3,
                            number_tasks_errors: 2,
                            number_tasks_submitted: 3,
                            response_data: {
                                geneData: {
                                    from_cache: false,
                                    response: {
                                        gene: {...}
                                    },
                                    status_code: 200,
                                    time_total: "00:00:00.35",
                                    time_request: "00:00:00.27",
                                    time_transfer: "00:00:00.08",
                                    url: "http://churchill-lab.jax.org/ensimpl/api/gene?id=ENSMUSG00000023224",
                                    url_id: "geneData"
                                },
                                expression: {
                                    from_cache: false,
                                    response: {
                                        ...
                                    },
                                    status_code: 404 or 500,    <-   ALSO AN ERROR
                                    time_total: "00:00:00.35",
                                    time_request: "00:00:00.27",
                                    time_transfer: "00:00:00.08",
                                    url: "http://myapiisawesome:8000/expression?dataset=dataset.islet.rnaseq&id=ENSMUSG00000023224",
                                    url_id: "expression"
                                },
                                lod: {
                                    error: "Connection Error"
                                    error_message: "HTTPConnectionPool(host='myapiisawesome', port=8000): Max retries exceeded with url: /lodscan?dataset=dataset.islet.rnaseq&nCores=2&id=ENSMUSG00000023224 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f861126ce48>: Failed to establish a new connection: [Errno -5] No address associated with hostname',))"
                                    url: "http://myapiisawesome:8000/lodscan?dataset=dataset.islet.rnaseq&nCores=2&id=ENSMUSG00000023224"
                                    url_id: "lod"
                                }
                            },
                            status: "DONE",
                            task_id: "d3c0b971-9794-4ed7-80cb-046710d2f9f5",
                            error: 'UNKNOWN ERROR'   <--- ONLY WHEN AN ERROR
                           }
                        */

                        if (data['status'] === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data['number_tasks_errors'] !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['error']}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data['number_tasks_errors'] === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['response']['error']}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the LOD SCAN.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 3 datasets to get

                                    // 1. gene information
                                    g.currentGene = data['response_data']['geneData']['response'];
                                    displayGeneData(g.currentGene.gene);
                                    let geneData = null;
                                    for (let key in g.currentGene.gene) {
                                        geneData = g.currentGene.gene[key];
                                    }

                                    let info = data['response_data']['geneData'];
                                    let dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Gene</strong></td><td>${geneData.id}</td></tr>
                                            </table>`;
                                    debugMessage('Gene Information', dbg);

                                    // 2. LOD score information
                                    plotLODChart(data['response_data']['lod']['response']['result']);

                                    info = data['response_data']['lod'];
                                    let len = info.response.result.length;

                                    dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>LOD Points</strong></td><td>${len}</td></tr>
                                            </table>`;
                                    debugMessage('LOD Information', dbg);

                                    // 3. Expression information
                                    g.expressionData = data['response_data']['expression']['response']['result'];

                                    if (g.dataSets[g.dataSetID].dataType === 'mRNA') {
                                        plotProfile(g.geneID);
                                    } else if (g.dataSets[g.dataSetID].dataType === 'protein') {
                                        plotProfile(g.proteinID);
                                    } else {
                                        console.log('ERROR');
                                    }


                                    info = data['response_data']['expression'];
                                    len = info.response.result.data.length;
                                    let dt = [];
                                    $.each(info.response.result.dataTypes, function (key, value) {
                                        dt.push(key)
                                    });

                                    dt = dt.join(',');
                                    dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Expression Points</strong></td><td>${len}</td></tr>
                                            <tr><td><strong>Data Types</strong></td><td>${dt}</td></tr>
                                            </table>`;
                                    debugMessage('Expression Information', dbg);

                                    // 4. Correlation
                                    info = data['response_data']['correlation'];
                                    len = info.response.result.length;

                                    displayCorrelation(info);

                                    dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Correlation Points</strong></td><td>${len}</td></tr>
                                            </table>`;
                                    debugMessage('Correlation Information', dbg);


                                    // first time we need to hide jumbotron instructions
                                    // it's ok to keep hiding
                                    $('#divWelcomeMesage').html('');
                                    $('#divLOD').removeClass("invisible");
                                    $('#divItemInfo').removeClass("invisible");
                                    $('#divSecondRow').removeClass("invisible");
                                    stopTask();
                                }
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function () {
                                updateGeneSelect(groupID);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=True) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                });
            }
        }

        /**
         * Perform gene search.
         */
        function selectPhenotype(phenotypeID, dataSetID) {
            //let urlGeneData = `http://churchill-lab.jax.org/ensimpl/api/gene?id=${geneID}`;
            let urlExpression = `{{ config.API_R_BASE }}/expression?dataset=${dataSetID}&id=${phenotypeID}`;
            let urlLOD = `{{ config.API_R_BASE }}/lodscan?dataset=${dataSetID}&id=${phenotypeID}`;
            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${dataSetID}&id=${phenotypeID}`;

            // TODO: make reset function
            g.currentGene = null;
            g.geneID = null;
            g.proteinID = null;
            g.phenotypeID = phenotypeID;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;
            g.correlationData = null;
            g.correlationDatasetID = dataSetID;

            // TODO: add ncores?, regress_local? to urlLOD, or handle backend?
            //URL += '&regress_local=' + regressLocal;
            //URL += '&ncores=' + _G.api_cores;
            {% if config.API_R_NUM_CORES %}
            urlLOD += '&nCores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'expression',
                    url: urlExpression
                }, {
                    url_id: 'correlation',
                    url: urlCorrelation
                }, {
                    url_id: 'lod',
                    url: urlLOD
                }]};

            startTask();

            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');

            g.chartLOD = null;
            $('#lodPlotChart').html('');


            g.chartCorrelation = null;
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            $('#correlationDatasetSelect').val(g.dataSetID);

            g.fact = {};
            $('#fact-svg-chart').html('');

            resetSecondaryPlots();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    g.runningTask = true;
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    //let status_url = request.getResponseHeader('Location');
                    //console.log('status_url=', status_url);
                    console.log('data.group_id=', data.group_id);
                    updatePhenotypeSelect(data.group_id);
                },
                error: function() {
                    alert('Unexpected error');
                }
            });
        }

        function showErrorMessage(message, details) {
            let errorMessage = `<div class="row text-left"><div class="col"><span class="font-weight-bold text-danger">${message}</span></div></div>`;

            if (details != null) {
                errorMessage += `<div class="row row-spacer"></div>
                                 <div class="row text-left">
                                     <div class="col">
                                         <div id="errorDetailAccordion" data-children=".item">
                                             <div class="item">
                                                 <a class="text-danger" data-toggle="collapse" data-parent="#errorDetailAccordion" href="#errorDetailAccordion1" role="button" aria-expanded="true" aria-controls="errorDetailAccordion1">
                                                     Details
                                                 </a>
                                                 <div id="errorDetailAccordion1" class="collapse" role="tabpanel">
                                                     <p class="mb-3">
                                                         ${details}
                                                     </p>
                                                 </div>
                                             </div>
                                         </div>
                                     </div>
                                 </div>`;
            }

            swal({
                  allowEnterKey: false,
                  allowEscapeKey: false,
                  allowOutsideClick: false,
                  buttonsStyling: false,
                  html: errorMessage,
                  showCancelButton: false,
                  showCloseButton: true,
                  showConfirmButton: false,
                  title: '<i class="fas fa-exclamation-circle"></i> Error!',
            });
        }


        function updatePhenotypeSelect(groupID) {
            // send GET request to status URL
            console.log('update progress, calling .ajax');

            console.log(groupID);

            if (g.runningTask) {
                let statusURL = `{{ url_for('api.api_status', group_id='', _external=True) }}${groupID}`;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    success: function (data, status, request) {
                        console.log('DATA=======', data);

                        /* A response will be like this:

                           {
                            number_tasks_completed: 3,
                            number_tasks_errors: 2,
                            number_tasks_submitted: 3,
                            response_data: {
                                expression: {
                                    from_cache: false,
                                    response: {
                                        ...
                                    },
                                    status_code: 404 or 500,    <-   ALSO AN ERROR
                                    time_total: "00:00:00.35",
                                    time_request: "00:00:00.27",
                                    time_transfer: "00:00:00.08",
                                    url: "http://myapiisawesome:8000/expression?dataset=dataset.islet.rnaseq&id=ENSMUSG00000023224",
                                    url_id: "expression"
                                },
                                lod: {
                                    error: "Connection Error"
                                    error_message: "HTTPConnectionPool(host='myapiisawesome', port=8000): Max retries exceeded with url: /lodscan?dataset=dataset.islet.rnaseq&nCores=2&id=ENSMUSG00000023224 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f861126ce48>: Failed to establish a new connection: [Errno -5] No address associated with hostname',))"
                                    url: "http://myapiisawesome:8000/lodscan?dataset=dataset.islet.rnaseq&nCores=2&id=ENSMUSG00000023224"
                                    url_id: "lod"
                                }
                            },
                            status: "DONE",
                            task_id: "d3c0b971-9794-4ed7-80cb-046710d2f9f5",
                            error: 'UNKNOWN ERROR'   <--- ONLY WHEN AN ERROR
                           }
                        */

                        if (data['status'] === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data['number_tasks_errors'] !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['error']}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data['number_tasks_errors'] === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data['response_data'], function (key, value) {
                                    if (value['status_code'] !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value['response']['error']}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the LOD SCAN.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 2 datasets to get
                                    displayPhenoData(g.phenotypeID);

                                    // 1. LOD score information
                                    plotLODChart(data['response_data']['lod']['response']['result']);

                                    let info = data['response_data']['lod'];
                                    let len = info.response.result.length;

                                    let dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>LOD Points</strong></td><td>${len}</td></tr>
                                            </table>`;
                                    debugMessage('LOD Information', dbg);

                                    // 2. Correlation
                                    info = data['response_data']['correlation'];
                                    len = info.response.result.length;

                                    displayCorrelation(info);

                                    dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Correlation Points</strong></td><td>${len}</td></tr>
                                            </table>`;
                                    debugMessage('Correlation Information', dbg);


                                    // 3. Expression information
                                    g.expressionData = data['response_data']['expression']['response']['result'];
                                    plotProfile(g.phenotypeID);

                                    info = data['response_data']['expression'];
                                    len = info.response.result.data.length;
                                    let dt = [];
                                    $.each(info.response.result.dataTypes, function (key, value) {
                                        dt.push(key)
                                    });

                                    dt = dt.join(',');
                                    dbg = `<table class="table table-sm">
                                            <tr><td><strong>Request Time</strong></td><td>${info.time_total}</td></tr>
                                            <tr><td><strong>API Time</strong></td><td>${info.response.time}</td></tr>
                                            <tr><td><strong>Cached</strong></td><td>${info.from_cache}</td></tr>
                                            <tr><td><strong>Expression Points</strong></td><td>${len}</td></tr>
                                            <tr><td><strong>Data Types</strong></td><td>${dt}</td></tr>
                                            </table>`;
                                    debugMessage('Expression Information', dbg);

                                    // first time we need to hide jumbotron instructions
                                    // it's ok to keep hiding
                                    $('#divWelcomeMesage').html('');
                                    $('#divLOD').removeClass("invisible");
                                    $('#divItemInfo').removeClass("invisible");
                                    $('#divSecondRow').removeClass("invisible");
                                    stopTask();
                                }
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            console.log('Not done, keep checking...');
                            setTimeout(function () {
                                updatePhenotypeSelect(groupID);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                console.log('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=True) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    console.log(data);
                });
            }
        }

        /**
         * Perform gene search.
         */
        function selectGeneProtein(geneID, proteinID, dataSetID) {
            let urlGeneData = `http://churchill-lab.jax.org/ensimpl/api/gene?id=${geneID}&species=${g.speciesID}&version=${g.ensemblVersion}`;
            let urlExpression = `{{ config.API_R_BASE }}/expression?dataset=${dataSetID}&id=`;
            let urlLOD = `{{ config.API_R_BASE }}/lodscan?dataset=${dataSetID}&id=`;
            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${dataSetID}&id=`;

            g.currentGene = null;
            g.geneID = geneID;
            g.proteinID = proteinID;
            g.phenotypeID = null;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;
            g.correlationData = null;
            g.correlationDatasetID = dataSetID;

            if ((proteinID === undefined) || (proteinID == null)) {
                urlExpression += geneID;
                urlLOD += geneID;
                urlCorrelation += geneID;
            } else {
                urlExpression += proteinID;
                urlLOD += proteinID;
                urlCorrelation += proteinID;
            }

            // TODO: add ncores?, regress_local? to urlLOD, or handle backend?
            //URL += '&regress_local=' + regressLocal;
            //URL += '&ncores=' + _G.api_cores;
            {% if config.API_R_NUM_CORES %}
            urlLOD += '&nCores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'geneData',
                    url: urlGeneData
                }, {
                    url_id: 'expression',
                    url: urlExpression
                }, {
                    url_id: 'correlation',
                    url: urlCorrelation
                }, {
                    url_id: 'lod',
                    url: urlLOD
                }]};

            startTask();

            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');

            g.chartLOD = null;
            $('#lodPlotChart').html('');

            g.chartCorrelation = null;
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            $('#correlationDatasetSelect').val(g.dataSetID);


            g.fact = {};
            $('#fact-svg-chart').html('');

            resetSecondaryPlots();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=True) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                success: function(data, status, request) {
                    g.runningTask = true;
                    console.log('data=', data);
                    console.log('status=', status);
                    console.log('request=', request);
                    //let status_url = request.getResponseHeader('Location');
                    //console.log('status_url=', status_url);
                    console.log('data.group_id=', data.group_id);
                    updateGeneSelect(data.group_id);
                },
                error: function() {
                    alert('Unexpected error');
                }
            });
        }

        return {
            g: g,
            init: init,
            findGene: findGene,
            startTask: startTask,
            stopTask: stopTask
        }
    }();

    $().ready(function () {
        QTL.init();

        // handle search
        $('#searchTerm').keypress(function(evt) {
            let code = evt.which ? evt.which : evt.keyCode;
            if (code === 13) {
                $(this).blur();
                $('#btnGo').focus().click();
            }
        });

        $('#btnGo').button().on('click', function(evt) {
            evt.preventDefault();
            let term = $('#searchTerm');
            QTL.findGene(term.val().trim().toUpperCase());
            return false;
        });


        $('#performBLUP').bootstrapSwitch();

        // set some global options for Highcharts
        Highcharts.setOptions({
            chart: {
                style: {
                    fontFamily: $('body').css('font-family')
                },
            },

            credits: {
                enabled: false
            },

            exporting: {
                buttons: {
                    contextButton: {
                        menuItems: [
                            'downloadPNG',
                            'downloadJPEG',
                            'downloadPDF',
                            'downloadSVG'
                        ]
                    }
                },
                //fallbackToExportServer: false,
            },
        });
    });
    </script>

{% endblock %}